<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Import | R community analysis</title>
  <meta name="description" content="NEOF book for the R community analysis workflow" />
  <meta name="generator" content="bookdown 0.31 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Import | R community analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the R community analysis workflow" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Import | R community analysis" />
  
  <meta name="twitter:description" content="NEOF book for the R community analysis workflow" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew R. Gemmell" />


<meta name="date" content="2022-12-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="03-R_packages.html"/>
<link rel="next" href="05-Preprocess_data.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/neof_small_logo.png"></a></li>
<li><a>R community analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="01-R_community_main_workflow.html"><a href="01-R_community_main_workflow.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html"><i class="fa fa-check"></i><b>2</b> Dataset &amp; workflow</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#dataset"><i class="fa fa-check"></i><b>2.1</b> Dataset</a></li>
<li class="chapter" data-level="2.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#workflow"><i class="fa fa-check"></i><b>2.2</b> Workflow</a></li>
<li class="chapter" data-level="2.3" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#scripts-and-working-directory"><i class="fa fa-check"></i><b>2.3</b> Scripts and working directory</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-R_packages.html"><a href="03-R_packages.html"><i class="fa fa-check"></i><b>3</b> R Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-R_packages.html"><a href="03-R_packages.html#r-packageslibraries"><i class="fa fa-check"></i><b>3.1</b> R packages/libraries</a></li>
<li class="chapter" data-level="3.2" data-path="03-R_packages.html"><a href="03-R_packages.html#the-grammar-of-graphics"><i class="fa fa-check"></i><b>3.2</b> The grammar of graphics</a></li>
<li class="chapter" data-level="3.3" data-path="03-R_packages.html"><a href="03-R_packages.html#phyloseq"><i class="fa fa-check"></i><b>3.3</b> phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Import.html"><a href="04-Import.html"><i class="fa fa-check"></i><b>4</b> Import</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Import.html"><a href="04-Import.html#script"><i class="fa fa-check"></i><b>4.1</b> Script</a></li>
<li class="chapter" data-level="4.2" data-path="04-Import.html"><a href="04-Import.html#qiime2r"><i class="fa fa-check"></i><b>4.2</b> qiime2R</a></li>
<li class="chapter" data-level="4.3" data-path="04-Import.html"><a href="04-Import.html#summarise-phyloseq"><i class="fa fa-check"></i><b>4.3</b> Summarise phyloseq</a></li>
<li class="chapter" data-level="4.4" data-path="04-Import.html"><a href="04-Import.html#save-the-phyloseq-object"><i class="fa fa-check"></i><b>4.4</b> Save the phyloseq object</a></li>
<li class="chapter" data-level="4.5" data-path="04-Import.html"><a href="04-Import.html#recap"><i class="fa fa-check"></i><b>4.5</b> Recap</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Preprocess_data.html"><a href="05-Preprocess_data.html"><i class="fa fa-check"></i><b>5</b> Transformations and preprocessing</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Preprocess_data.html"><a href="05-Preprocess_data.html#script-setup"><i class="fa fa-check"></i><b>5.1</b> Script setup</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="06-Taxonomy.html"><a href="06-Taxonomy.html"><i class="fa fa-check"></i><b>6</b> Taxonomy</a></li>
<li class="chapter" data-level="7" data-path="07-Alpha.html"><a href="07-Alpha.html"><i class="fa fa-check"></i><b>7</b> Alpha</a></li>
<li class="chapter" data-level="8" data-path="08-Beta.html"><a href="08-Beta.html"><i class="fa fa-check"></i><b>8</b> Beta</a></li>
<li class="chapter" data-level="9" data-path="09-Differential_abundance_analysis.html"><a href="09-Differential_abundance_analysis.html"><i class="fa fa-check"></i><b>9</b> Differential abundance analysis</a></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="10-Appendix.html"><a href="10-Appendix.html"><i class="fa fa-check"></i><b>A</b> Installing required packages</a></li>
<li class="chapter" data-level="B" data-path="10-Appendix.html"><a href="10-Appendix.html#ggplot2_appendix"><i class="fa fa-check"></i><b>B</b> ggplot2</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R community analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="import" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Import<a href="04-Import.html#import" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><img src="figures/import.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Before carrying out any analysis we first need to import our QIIME2 artifacts into a phyloseq object. Thankfully there is an R package called <code>qiime2R</code></p>
<div id="script" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Script<a href="04-Import.html#script" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/r_script.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Prior to any coding, open and save a new script called "01-Import.R". We will be creating a new script for each chapter and numbering them so we can easily see the order of scripts.</p>
</div>
<div id="qiime2r" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> qiime2R<a href="04-Import.html#qiime2r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/qiime2r.png" width="40%" style="display: block; margin: auto;" /></p>
<p><a href="https://github.com/jbisanz/qiime2R"><code>qiime2R</code></a> is an R package for importing QIIME2 artifacts into a R phyloseq object. The package contains many different commands. Its function <code>read_qza()</code> can read a single artifact at a time.</p>
<p>The best way to import all your QIIME2 artifacts is with the <code>qza_to_phyloseq()</code> function. In your "01-Import.R" script, add the following and run the commands.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="04-Import.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Set you working directory</span></span>
<span id="cb5-2"><a href="04-Import.html#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#This is useful to have at the top of your scripts</span></span>
<span id="cb5-3"><a href="04-Import.html#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Ensure you set it as your analysis directory</span></span>
<span id="cb5-4"><a href="04-Import.html#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;~/R/R_community_course/freshwater_data_analysis/&quot;</span>)</span>
<span id="cb5-5"><a href="04-Import.html#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="04-Import.html#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Load the package/library</span></span>
<span id="cb5-7"><a href="04-Import.html#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;qiime2R&quot;</span>)</span>
<span id="cb5-8"><a href="04-Import.html#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="04-Import.html#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Import data</span></span>
<span id="cb5-10"><a href="04-Import.html#cb5-10" aria-hidden="true" tabindex="-1"></a>pseq <span class="ot">&lt;-</span> qiime2R<span class="sc">::</span><span class="fu">qza_to_phyloseq</span>(</span>
<span id="cb5-11"><a href="04-Import.html#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="st">&quot;table-dada2.qza&quot;</span>,</span>
<span id="cb5-12"><a href="04-Import.html#cb5-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">tree =</span> <span class="st">&quot;rooted-tree.qza&quot;</span>,</span>
<span id="cb5-13"><a href="04-Import.html#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">taxonomy =</span> <span class="st">&quot;taxonomy.sklearn.qza&quot;</span>,</span>
<span id="cb5-14"><a href="04-Import.html#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> <span class="st">&quot;media_metadata.txt&quot;</span></span>
<span id="cb5-15"><a href="04-Import.html#cb5-15" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>This command creates a phyloseq object named <code>pseq</code>. It contains:</p>
<ul>
<li>The ASV abundance table (<code>features = "table-dada2.qza"</code>).</li>
<li>The rooted phylogenetic tree (<code>tree = "rooted-tree.qza"</code>).</li>
<li>The taxonomic classifications of the ASVs (<code>taxonomy = "taxonomy.sklearn.qza"</code>).</li>
<li>The sample metadata (<code>metadata = "media_metadata.txt"</code>)</li>
</ul>
</div>
<div id="summarise-phyloseq" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Summarise phyloseq<a href="04-Import.html#summarise-phyloseq" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/list_blue.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Now that we have imported the data we can extract some summary information from it.</p>
<p>First we will use the <code>microbiome</code> package with its <code>summarize_phyloseq()</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="04-Import.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Load microbiome library</span></span>
<span id="cb6-2"><a href="04-Import.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb6-3"><a href="04-Import.html#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Summary of phyloseq object</span></span>
<span id="cb6-4"><a href="04-Import.html#cb6-4" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq)</span></code></pre></div>
<p>This gives us a plethora of information:</p>
<ul>
<li>The top line tells us if the data is compositional (relative abundance).</li>
<li>We get the following list of values in a paragraph and via a list.
<ul>
<li>Min. number of reads: Number of reads in the sample with the lowest number of reads.</li>
<li>Max. number of reads: Number of reads in the sample with the largest number of reads.</li>
<li>Total number of reads: Sum of all reads across all samples.</li>
<li>Average number of reads: Sum of all reads / number of samples.</li>
<li>Median number of reads: Midpoint read abundance across samples.</li>
<li>Sparsity: See expandable box further down.</li>
<li>Any OTU sum to 1 or less?: States if there are any ASVs with a summed abundance of 1 or less across all the samples.</li>
<li>Number of singletons: Number of ASVs with a sum of 1 or less across all samples.</li>
<li>Percent of OTUs that are singletons: Percentage of ASVs that only contain one read across all the samples.</li>
<li>Number of sample variables are: Number of sample variables/groupings in our metadata.</li>
<li>The last line shows the names of the sample variables/groupings in our metadata.</li>
</ul></li>
</ul>
<div class="webex-solution">
<button>
Sparsity
</button>
<p>Sparsity is a measure of the number of 0s in a table. It can be represented by the following equation:</p>
<p><span class="math display">\[
sparsity = Z/C
\]</span></p>
<p>Where:</p>
<ul>
<li>Z = The number of cells that equal zero.</li>
<li>C = The total number of cells.</li>
</ul>
<p>Let's look at an example of an abundance table with a small amount of ASVs and Samples.</p>
<table>
<thead>
<tr class="header">
<th></th>
<th>Sample1</th>
<th>Sample2</th>
<th>Sample3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ASV1</td>
<td>0</td>
<td>10</td>
<td>24</td>
</tr>
<tr class="even">
<td>ASV2</td>
<td>1</td>
<td>0</td>
<td>37</td>
</tr>
<tr class="odd">
<td>ASV3</td>
<td>6</td>
<td>25</td>
<td>0</td>
</tr>
<tr class="even">
<td>ASV4</td>
<td>51</td>
<td>2</td>
<td>0</td>
</tr>
</tbody>
</table>
<ul>
<li>This abundance table has 12 cells, 3 samples * 4 ASVs.</li>
<li>Of these cells, 4 have an abundance of zero.</li>
<li>4/12 = 0.3333, therefore its sparsity is 0.3333.</li>
</ul>
<p>Sparsity can be any value from 0-1. The higher the value the more sparse it is, with a value of 1 meaning all the cells have an abundance of zero. The lower the value the less sparse it is, with a value of 0 meaning all the cells have an abundance of 1 or more.</p>
<p>16S data is known to be sparse so high sparsity is not unexpected. Keep in mind that lower levels of taxa (ASVs, Species, &amp; Genera) will generally have more sparse tables that higher levels of taxa (Kingdom, Phylum, Class).</p>
</div>
<p>If you would like to see how the function calculates its values you can highlight the function on the script editor and press "F2".</p>
</div>
<div id="save-the-phyloseq-object" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Save the phyloseq object<a href="04-Import.html#save-the-phyloseq-object" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/r_save.png" width="15%" style="display: block; margin: auto;" /></p>
<p>When using multiple scripts for analysis it is useful to save the R objects that will be used in different script. This can be carried out with the function <code>save()</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="04-Import.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(pseq, <span class="at">file =</span> <span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
<p>This saves our object <code>pseq</code> into the file <code>phyloseq.RData</code>. The suffix <code>R.Data</code> is the normal convention.</p>
<p>Since our final object is saved we no longer need the objects in the environment. Therefore, you can use the sweep button in the Environment pane to remove them.</p>
<p><img src="figures/RStudio_env_sweep.png" width="40%" style="display: block; margin: auto;" /></p>
</div>
<div id="recap" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> Recap<a href="04-Import.html#recap" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/recap.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have imported our QIIME2 artifacts into one phyloseq object so we can analyse the data in R. This object has been saved into a ".RData" file which we will load in the next chapter.</p>

</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  if (t = document.getElementById("webex-total_correct")) {
    var correct = document.getElementsByClassName("webex-correct").length;
    var solvemes = document.getElementsByClassName("webex-solveme").length;
    var radiogroups = document.getElementsByClassName("webex-radiogroup").length;
    var selects = document.getElementsByClassName("webex-select").length;
    
    t.innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");
  
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");
  
  var cl = this.classList
  
  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;
  
  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }
  
  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

window.onload = function() {
  console.log("onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }
  
  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="03-R_packages.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="05-Preprocess_data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
