<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Minimum read depth | R community analysis</title>
  <meta name="description" content="NEOF book for the R community analysis workflow" />
  <meta name="generator" content="bookdown 0.31 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Minimum read depth | R community analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the R community analysis workflow" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Minimum read depth | R community analysis" />
  
  <meta name="twitter:description" content="NEOF book for the R community analysis workflow" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew R. Gemmell" />


<meta name="date" content="2023-03-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="08-Summarise_phyloseq.html"/>
<link rel="next" href="10-Relative_abundance.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/neof_small_logo.png"></a></li>
<li><a>R community analysis</a></li>

<li class="divider"></li>
<li class="part"><span><b>Intro</b></span></li>
<li class="chapter" data-level="1" data-path="01-R_community_main_workflow.html"><a href="01-R_community_main_workflow.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html"><i class="fa fa-check"></i><b>2</b> Dataset &amp; workflow</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#dataset"><i class="fa fa-check"></i><b>2.1</b> Dataset</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sites"><i class="fa fa-check"></i><b>2.1.1</b> Sites</a></li>
<li class="chapter" data-level="2.1.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#culture-media"><i class="fa fa-check"></i><b>2.1.2</b> Culture media</a></li>
<li class="chapter" data-level="2.1.3" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#summary-questions"><i class="fa fa-check"></i><b>2.1.3</b> Summary &amp; questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#workflow"><i class="fa fa-check"></i><b>2.2</b> Workflow</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-R_packages.html"><a href="03-R_packages.html"><i class="fa fa-check"></i><b>3</b> R Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-R_packages.html"><a href="03-R_packages.html#r-packageslibraries"><i class="fa fa-check"></i><b>3.1</b> R packages/libraries</a></li>
<li class="chapter" data-level="3.2" data-path="03-R_packages.html"><a href="03-R_packages.html#the-grammar-of-graphics"><i class="fa fa-check"></i><b>3.2</b> The grammar of graphics</a></li>
<li class="chapter" data-level="3.3" data-path="03-R_packages.html"><a href="03-R_packages.html#phyloseq"><i class="fa fa-check"></i><b>3.3</b> phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Setup.html"><a href="04-Setup.html"><i class="fa fa-check"></i><b>4</b> Set-up</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Setup.html"><a href="04-Setup.html#cluster"><i class="fa fa-check"></i><b>4.1</b> Logon instructions</a></li>
<li class="chapter" data-level="4.2" data-path="04-Setup.html"><a href="04-Setup.html#mamba"><i class="fa fa-check"></i><b>4.2</b> Mamba</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Jupyter.html"><a href="05-Jupyter.html"><i class="fa fa-check"></i><b>5</b> Jupyter</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Jupyter.html"><a href="05-Jupyter.html#open-jupyter-notebook"><i class="fa fa-check"></i><b>5.1</b> Open Jupyter-notebook</a></li>
<li class="chapter" data-level="5.2" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-r-notebook"><i class="fa fa-check"></i><b>5.2</b> Create R notebook</a></li>
<li class="chapter" data-level="5.3" data-path="05-Jupyter.html"><a href="05-Jupyter.html#cells-and-code"><i class="fa fa-check"></i><b>5.3</b> Cells and code</a></li>
<li class="chapter" data-level="5.4" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-new-cells"><i class="fa fa-check"></i><b>5.4</b> Create new cells</a></li>
<li class="chapter" data-level="5.5" data-path="05-Jupyter.html"><a href="05-Jupyter.html#running-code"><i class="fa fa-check"></i><b>5.5</b> Running code</a></li>
<li class="chapter" data-level="5.6" data-path="05-Jupyter.html"><a href="05-Jupyter.html#saving-the-file"><i class="fa fa-check"></i><b>5.6</b> Saving the file</a></li>
<li class="chapter" data-level="5.7" data-path="05-Jupyter.html"><a href="05-Jupyter.html#title-cells-with-markdown"><i class="fa fa-check"></i><b>5.7</b> Title cells with markdown</a></li>
<li class="chapter" data-level="5.8" data-path="05-Jupyter.html"><a href="05-Jupyter.html#close-the-notebook"><i class="fa fa-check"></i><b>5.8</b> Close the notebook</a></li>
</ul></li>
<li class="part"><span><b>Data preparation</b></span></li>
<li class="chapter" data-level="6" data-path="06-Preprocess_part.html"><a href="06-Preprocess_part.html"><i class="fa fa-check"></i><b>6</b> Data prep intro</a></li>
<li class="chapter" data-level="7" data-path="07-Import.html"><a href="07-Import.html"><i class="fa fa-check"></i><b>7</b> Import</a>
<ul>
<li class="chapter" data-level="7.1" data-path="07-Import.html"><a href="07-Import.html#import-notebook"><i class="fa fa-check"></i><b>7.1</b> Import: notebook</a></li>
<li class="chapter" data-level="7.2" data-path="07-Import.html"><a href="07-Import.html#qiime2r"><i class="fa fa-check"></i><b>7.2</b> qiime2R</a></li>
<li class="chapter" data-level="7.3" data-path="07-Import.html"><a href="07-Import.html#import-summarise-phyloseq"><i class="fa fa-check"></i><b>7.3</b> Import: Summarise phyloseq</a></li>
<li class="chapter" data-level="7.4" data-path="07-Import.html"><a href="07-Import.html#save-the-phyloseq-object"><i class="fa fa-check"></i><b>7.4</b> Save the phyloseq object</a></li>
<li class="chapter" data-level="7.5" data-path="07-Import.html"><a href="07-Import.html#import-recap"><i class="fa fa-check"></i><b>7.5</b> Import: recap</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html"><i class="fa fa-check"></i><b>8</b> Summarise phyloseq</a>
<ul>
<li class="chapter" data-level="8.1" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#setup"><i class="fa fa-check"></i><b>8.1</b> Setup</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-header"><i class="fa fa-check"></i><b>8.1.1</b> Summarise header</a></li>
<li class="chapter" data-level="8.1.2" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-phyloseq"><i class="fa fa-check"></i><b>8.1.2</b> Summarise phyloseq</a></li>
<li class="chapter" data-level="8.1.3" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#reads-per-sample"><i class="fa fa-check"></i><b>8.1.3</b> Reads per sample</a></li>
<li class="chapter" data-level="8.1.4" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#asvs-per-sample"><i class="fa fa-check"></i><b>8.1.4</b> ASVs per sample</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-recap"><i class="fa fa-check"></i><b>8.2</b> Summarise: recap</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html"><i class="fa fa-check"></i><b>9</b> Minimum read depth</a>
<ul>
<li class="chapter" data-level="9.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#considerations"><i class="fa fa-check"></i><b>9.1</b> Considerations</a></li>
<li class="chapter" data-level="9.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#minimum-read-depth-section-title"><i class="fa fa-check"></i><b>9.2</b> Minimum read depth section title</a></li>
<li class="chapter" data-level="9.3" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#read-depth-vector-and-histogram"><i class="fa fa-check"></i><b>9.3</b> Read depth vector and histogram</a></li>
<li class="chapter" data-level="9.4" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#sample-depth-boxplot"><i class="fa fa-check"></i><b>9.4</b> Sample depth boxplot</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#creating-data-frame-for-boxplots"><i class="fa fa-check"></i><b>9.4.1</b> Creating data frame for boxplots</a></li>
<li class="chapter" data-level="9.4.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#depth-boxplots"><i class="fa fa-check"></i><b>9.4.2</b> Depth boxplots</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#rarefaction-curve"><i class="fa fa-check"></i><b>9.5</b> Rarefaction curve</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html"><i class="fa fa-check"></i><b>10</b> Relative abundance</a></li>
<li class="chapter" data-level="11" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html"><i class="fa fa-check"></i><b>11</b> Rarefaction</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R community analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="minimum-read-depth" class="section level1 hasAnchor" number="9">
<h1><span class="header-section-number">Chapter 9</span> Minimum read depth<a href="09-Minimum_read_depth.html#minimum-read-depth" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p><img src="figures/depth_diver.png" width="20%" style="display: block; margin: auto;" /></p>
<p>It is good to remove samples with a very low read depth (number of sequencing reads).
However, it is not trivial to determine what an appropriate read depth is.
This value will vary from study to study.
Normally, for 16S data, a depth of at least 20K per sample is suggested. However, this is the general consensus for human microbiome data.</p>
<p>In this chapter we will:</p>
<ul>
<li>Cover a brief intro to considertions of what is an acceptable minimum depth for your dataset.</li>
<li>Reinvestigate our sample depths with a previously created histogram.</li>
<li>View the depth ranges of different sample groups (site &amp; media) with box plots.</li>
<li>Create a rarefaction curve to assess if the depth of our samples have captured a good amount of biodiversity.</li>
<li>Demonstrate how to filter samples by depth.</li>
</ul>
<div id="considerations" class="section level2 hasAnchor" number="9.1">
<h2><span class="header-section-number">9.1</span> Considerations<a href="09-Minimum_read_depth.html#considerations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/consider.png" width="15%" style="display: block; margin: auto;" /></p>
<p>There are 3 main considerations to take into account for what is an appropriate depth for your dataset.</p>
<ul>
<li><strong>The biodiversity of your samples.</strong>
<ul>
<li>If your sample is very biodiverse, such as the human gut microbiome, you will need a good depth (&gt;20K per sample).</li>
<li>If your sample is less biodiverse, such as many geological environments or skin, then you will not need as much read depth.</li>
<li>Rarefaction curves are a good method to determine if your samples have enough depth. We will look at this in this chapter.</li>
</ul></li>
<li><strong>The biomass of your samples.</strong>
<ul>
<li>Some environments are hard to extract DNA from.</li>
<li>If this is the case for you, then people will hopefully accept that this is an unfortunate reality of life and you will use what you can.</li>
<li>However, be careful of your conclusions, if you think your data doesn't have as much as it could do not make very definitive detailed claims.</li>
</ul></li>
<li><strong>Read depth of sample groups.</strong>
<ul>
<li>It may be possible that a few samples have a much lower depth than the rest and so you may think to remove them.</li>
<li>However, these may all come from the same sample group and so you have lost all information of one group.</li>
<li>For instance, you may be comparing different geological surfaces and your rock samples have much lower read depths than the various soil samples.</li>
<li>For comparisons including the lower depth sample group (e.g. rock samples) you will need to retrain the lower depth sample.</li>
</ul></li>
</ul>
<p>That is a brief overview of that topic. If you are interested in more I suggest you look at papers where they have studied an environment similar to yours.</p>
</div>
<div id="minimum-read-depth-section-title" class="section level2 hasAnchor" number="9.2">
<h2><span class="header-section-number">9.2</span> Minimum read depth section title<a href="09-Minimum_read_depth.html#minimum-read-depth-section-title" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/header_l2.png" width="10%" style="display: block; margin: auto;" /></p>
<p>In the next section of our <code>jupyter-notebook</code> we will investigate what the minimum read depth should be and remove sample below this.</p>
<p>Create a new <strong>"Markdown"</strong> cell and add the following 2nd level header.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="09-Minimum_read_depth.html#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Minimum read depth</span></span></code></pre></div>
</div>
<div id="read-depth-vector-and-histogram" class="section level2 hasAnchor" number="9.3">
<h2><span class="header-section-number">9.3</span> Read depth vector and histogram<a href="09-Minimum_read_depth.html#read-depth-vector-and-histogram" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/histogram.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We have already created a vector and histogram of the read depths across our samples. Scroll up your notebook to view these and answer the following MCQ:</p>
<ul>
<li>What is the approximate read depth range of our dataset?
<div id="radio_DAGCSXEHRD" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_DAGCSXEHRD" value=""></input> <span><strong>1-8</strong></span></label><label><input type="radio" autocomplete="off" name="radio_DAGCSXEHRD" value="answer"></input> <span><strong>10,000-18,000</strong></span></label><label><input type="radio" autocomplete="off" name="radio_DAGCSXEHRD" value=""></input> <span><strong>20,000+</strong></span></label>
</div></li>
</ul>
<p>This is a good first step but what if we want to know how the read depths vary between sample groups?</p>
</div>
<div id="sample-depth-boxplot" class="section level2 hasAnchor" number="9.4">
<h2><span class="header-section-number">9.4</span> Sample depth boxplot<a href="09-Minimum_read_depth.html#sample-depth-boxplot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/boxplot.png" width="15%" style="display: block; margin: auto;" /></p>
<p>We are going to use <code>ggplot2</code> to create a couple boxplots to show the sequencing depth ranges of the different sample groups (Site &amp; Media).</p>
<div id="creating-data-frame-for-boxplots" class="section level3 hasAnchor" number="9.4.1">
<h3><span class="header-section-number">9.4.1</span> Creating data frame for boxplots<a href="09-Minimum_read_depth.html#creating-data-frame-for-boxplots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="figures/metadata.png" width="10%" style="display: block; margin: auto;" /></p>
<p>First, we need to create an object containing our sample names, Site &amp; Media information, and the depth.
We will use this object to produce our boxplots.
Thankfully the metadata in our <code>phyloseq</code> object contains all this information except the depth.</p>
<p>In a new cell write and run the below code. This will extract the sample data (metadata) to a new object and display the top 6 rows fo this new object.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="09-Minimum_read_depth.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract sample data as a separate R object</span></span>
<span id="cb25-2"><a href="09-Minimum_read_depth.html#cb25-2" aria-hidden="true" tabindex="-1"></a>abundance_metadf <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(pseq)</span>
<span id="cb25-3"><a href="09-Minimum_read_depth.html#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View top 6 rows of metadata data frame</span></span>
<span id="cb25-4"><a href="09-Minimum_read_depth.html#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundance_metadf)</span></code></pre></div>
<p>The 2 new functions above are:</p>
<ul>
<li><code>phyloseq::sample_data()</code>: Extracts the sample data (metadata) data frame from a <code>phyloseq</code> object.</li>
<li><code>head()</code>: Returns the first 6 parts of an R object by default.
<ul>
<li>It can be used for a vector, matrix, table, data frame, or function.</li>
<li>In the case of a data frame it returns the first 6 rows.</li>
</ul></li>
</ul>
<p>We need to add the depth information to our new data frame.
We have extracted this previously into an object called <code>sample_depths</code>.
However, before adding it we want to check it has the same order of samples as the rows in <code>abundance_metadf</code>.</p>
<p>Write and run the below script in the same cell. The code uses <code>head()</code> to view the first 6 elements of <code>sample_depths</code> and the row names of <code>abundance_metadf</code>. Then the function <code>idnetical()</code> is used to see if they are identical (<code>TRUE</code>) or not (<code>FALSE</code>).</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="09-Minimum_read_depth.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Check if our vector of sample_depths has the same order as our metadata rows</span></span>
<span id="cb26-2"><a href="09-Minimum_read_depth.html#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">names</span>(sample_depths))</span>
<span id="cb26-3"><a href="09-Minimum_read_depth.html#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">row.names</span>(abundance_metadf))</span>
<span id="cb26-4"><a href="09-Minimum_read_depth.html#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(<span class="fu">names</span>(sample_depths),<span class="fu">row.names</span>(abundance_metadf))</span></code></pre></div>
<p>The order of samples is identical so we can add the depth information to <code>abundance_metadf</code>. Carry this out in the same cell with the code below.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="09-Minimum_read_depth.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Add sample depths to metadata data frame</span></span>
<span id="cb27-2"><a href="09-Minimum_read_depth.html#cb27-2" aria-hidden="true" tabindex="-1"></a>abundance_metadf[,<span class="st">&quot;depth&quot;</span>] <span class="ot">&lt;-</span> sample_depths</span>
<span id="cb27-3"><a href="09-Minimum_read_depth.html#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co">#View top 6 rows of edited metadata dataframe</span></span>
<span id="cb27-4"><a href="09-Minimum_read_depth.html#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundance_metadf)</span></code></pre></div>
<p>Great! We will next use this data frame to create 2 boxplots.</p>
</div>
<div id="depth-boxplots" class="section level3 hasAnchor" number="9.4.2">
<h3><span class="header-section-number">9.4.2</span> Depth boxplots<a href="09-Minimum_read_depth.html#depth-boxplots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="figures/boxplots_groups.png" width="10%" style="display: block; margin: auto;" /></p>
<p>We are going to create 2 boxplots with <code>ggplot2</code>.
We won't go into too much detail on how the code works here, instead learning more later in this book.</p>
<p>The code below creates a <code>ggplot2</code> boxplot. We carry this out with 2 functions:</p>
<ul>
<li><code>ggplot()</code>: This creates a <code>ggplot2</code> object, storing the information and aesthetics.
<ul>
<li>The first option is the data we want to use for plotting (<code>abundance_metadf</code>).</li>
<li>The second option is the aesthetics (<code>aes()</code>) to plot. In this case we want the depth column to be plotted on the y-axis (<code>y=depth</code>) and the site column to be plotted on the x-axis (<code>x=site</code>).</li>
</ul></li>
<li><code>+</code>: We need to have a <code>+</code> at the end of the <code>ggplot()</code> function to add the next component of the plot.</li>
<li><code>geom_boxplot()</code>: This adds a layer to our <code>ggplot2</code> object. In this case it converts the <code>ggplot</code> object, which is just information, into a boxplot.</li>
</ul>
<p>Write and run the following code in a new cell:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="09-Minimum_read_depth.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create ggplot2 boxplot of depth by size</span></span>
<span id="cb28-2"><a href="09-Minimum_read_depth.html#cb28-2" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(abundance_metadf, <span class="at">aes=</span>(<span class="at">y=</span>depth, <span class="at">x=</span>site)) <span class="sc">+</span></span>
<span id="cb28-3"><a href="09-Minimum_read_depth.html#cb28-3" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span></code></pre></div>
<ul>
<li>Which site has the highest median depth (middle line of boxplot)?
<div id="radio_ZJVADIHEZB" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_ZJVADIHEZB" value="answer"></input> <span><strong>LD (Lower Durance)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZJVADIHEZB" value=""></input> <span><strong>MD (Middle Durance)</strong></span></label><label><input type="radio" autocomplete="off" name="radio_ZJVADIHEZB" value=""></input> <span><strong>UD (Upper Durance</strong></span></label>
</div></li>
</ul>
<p>We will use the same code to plot the depth by media.
You can copy and paste the code changing the x aesthetic to media (<code>x=media</code>).
Carry this out in the same cell and run the code.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="09-Minimum_read_depth.html#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create ggplot2 boxplot of depth by size</span></span>
<span id="cb29-2"><a href="09-Minimum_read_depth.html#cb29-2" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(abundance_metadf, <span class="at">aes=</span>(<span class="at">y=</span>depth, <span class="at">x=</span>media)) <span class="sc">+</span></span>
<span id="cb29-3"><a href="09-Minimum_read_depth.html#cb29-3" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span></code></pre></div>
<ul>
<li>Which media has the lowest median depth (middle line of boxplot)?
<div id="radio_LNZULRIFNH" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_LNZULRIFNH" value="answer"></input> <span><strong>CVP</strong></span></label><label><input type="radio" autocomplete="off" name="radio_LNZULRIFNH" value=""></input> <span><strong>ENV</strong></span></label><label><input type="radio" autocomplete="off" name="radio_LNZULRIFNH" value=""></input> <span><strong>TSA</strong></span></label>
</div></li>
</ul>
<p>From the boxplots we can see there is no drastic difference between the depths of the different sample groups. We will therefore continue and make some rarefaction curves to further assess the depth of our samples.</p>
<p>For more resources on <code>ggplot2</code> please see the <a href="#ggplot2_appendix">appendix</a> of this book.</p>
</div>
</div>
<div id="rarefaction-curve" class="section level2 hasAnchor" number="9.5">
<h2><span class="header-section-number">9.5</span> Rarefaction curve<a href="09-Minimum_read_depth.html#rarefaction-curve" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/rarefaction.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Our read depths appear a bit low.
However, this might be fine for our dataset since we are use surface water samples rather than human gut microbiome samples.
Let's see how our samples look with a rarefaction curve.</p>
<p>#Sample with too few samples?
#In your analysis you may have a sample with too few reads
#All our samples are fine for a tutorial case but let us say we only wanted to
#keep samples with &gt;11k reads
#We could remove samples with the following code
pseq_min11K &lt;- phyloseq::subset_samples(pseq, reads_sample &gt; 11000)
microbiome::summarize_phyloseq(pseq_min11K)
microbiome::readcount(pseq_min11K)</p>
<p>#We won't be using this as we are happy with our sample numbers
#Let us therefore remove this sample subsetted variable
rm(pseq_min11K)</p>

</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  if (t = document.getElementById("webex-total_correct")) {
    var correct = document.getElementsByClassName("webex-correct").length;
    var solvemes = document.getElementsByClassName("webex-solveme").length;
    var radiogroups = document.getElementsByClassName("webex-radiogroup").length;
    var selects = document.getElementsByClassName("webex-select").length;
    
    t.innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");
  
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");
  
  var cl = this.classList
  
  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;
  
  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }
  
  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

window.onload = function() {
  console.log("onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }
  
  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="08-Summarise_phyloseq.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="10-Relative_abundance.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
