<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Relative abundance | R community analysis</title>
  <meta name="description" content="NEOF book for the R community analysis workflow" />
  <meta name="generator" content="bookdown 0.31 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Relative abundance | R community analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the R community analysis workflow" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Relative abundance | R community analysis" />
  
  <meta name="twitter:description" content="NEOF book for the R community analysis workflow" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew R. Gemmell" />


<meta name="date" content="2023-03-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="09-Minimum_read_depth.html"/>
<link rel="next" href="11-Rarefaction.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/neof_small_logo.png"></a></li>
<li><a>R community analysis</a></li>

<li class="divider"></li>
<li class="part"><span><b>Intro</b></span></li>
<li class="chapter" data-level="1" data-path="01-R_community_main_workflow.html"><a href="01-R_community_main_workflow.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html"><i class="fa fa-check"></i><b>2</b> Dataset &amp; workflow</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#dataset"><i class="fa fa-check"></i><b>2.1</b> Dataset</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sites"><i class="fa fa-check"></i><b>2.1.1</b> Sites</a></li>
<li class="chapter" data-level="2.1.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#culture-media"><i class="fa fa-check"></i><b>2.1.2</b> Culture media</a></li>
<li class="chapter" data-level="2.1.3" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#summary-questions"><i class="fa fa-check"></i><b>2.1.3</b> Summary &amp; questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#workflow"><i class="fa fa-check"></i><b>2.2</b> Workflow</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-R_packages.html"><a href="03-R_packages.html"><i class="fa fa-check"></i><b>3</b> R Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-R_packages.html"><a href="03-R_packages.html#r-packageslibraries"><i class="fa fa-check"></i><b>3.1</b> R packages/libraries</a></li>
<li class="chapter" data-level="3.2" data-path="03-R_packages.html"><a href="03-R_packages.html#the-grammar-of-graphics"><i class="fa fa-check"></i><b>3.2</b> The grammar of graphics</a></li>
<li class="chapter" data-level="3.3" data-path="03-R_packages.html"><a href="03-R_packages.html#phyloseq"><i class="fa fa-check"></i><b>3.3</b> phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Setup.html"><a href="04-Setup.html"><i class="fa fa-check"></i><b>4</b> Set-up</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Setup.html"><a href="04-Setup.html#cluster"><i class="fa fa-check"></i><b>4.1</b> Logon instructions</a></li>
<li class="chapter" data-level="4.2" data-path="04-Setup.html"><a href="04-Setup.html#mamba"><i class="fa fa-check"></i><b>4.2</b> Mamba</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Jupyter.html"><a href="05-Jupyter.html"><i class="fa fa-check"></i><b>5</b> Jupyter</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Jupyter.html"><a href="05-Jupyter.html#open-jupyter-notebook"><i class="fa fa-check"></i><b>5.1</b> Open Jupyter-notebook</a></li>
<li class="chapter" data-level="5.2" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-r-notebook"><i class="fa fa-check"></i><b>5.2</b> Create R notebook</a></li>
<li class="chapter" data-level="5.3" data-path="05-Jupyter.html"><a href="05-Jupyter.html#cells-and-code"><i class="fa fa-check"></i><b>5.3</b> Cells and code</a></li>
<li class="chapter" data-level="5.4" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-new-cells"><i class="fa fa-check"></i><b>5.4</b> Create new cells</a></li>
<li class="chapter" data-level="5.5" data-path="05-Jupyter.html"><a href="05-Jupyter.html#running-code"><i class="fa fa-check"></i><b>5.5</b> Running code</a></li>
<li class="chapter" data-level="5.6" data-path="05-Jupyter.html"><a href="05-Jupyter.html#saving-the-file"><i class="fa fa-check"></i><b>5.6</b> Saving the file</a></li>
<li class="chapter" data-level="5.7" data-path="05-Jupyter.html"><a href="05-Jupyter.html#title-cells-with-markdown"><i class="fa fa-check"></i><b>5.7</b> Title cells with markdown</a></li>
<li class="chapter" data-level="5.8" data-path="05-Jupyter.html"><a href="05-Jupyter.html#close-the-notebook"><i class="fa fa-check"></i><b>5.8</b> Close the notebook</a></li>
<li class="chapter" data-level="5.9" data-path="05-Jupyter.html"><a href="05-Jupyter.html#video-tutorial"><i class="fa fa-check"></i><b>5.9</b> Video tutorial</a></li>
</ul></li>
<li class="part"><span><b>Data preparation</b></span></li>
<li class="chapter" data-level="6" data-path="06-Preprocess_part.html"><a href="06-Preprocess_part.html"><i class="fa fa-check"></i><b>6</b> Data prep intro</a></li>
<li class="chapter" data-level="7" data-path="07-Import.html"><a href="07-Import.html"><i class="fa fa-check"></i><b>7</b> Import</a>
<ul>
<li class="chapter" data-level="7.1" data-path="07-Import.html"><a href="07-Import.html#import-notebook"><i class="fa fa-check"></i><b>7.1</b> Import: notebook</a></li>
<li class="chapter" data-level="7.2" data-path="07-Import.html"><a href="07-Import.html#qiime2r"><i class="fa fa-check"></i><b>7.2</b> qiime2R</a></li>
<li class="chapter" data-level="7.3" data-path="07-Import.html"><a href="07-Import.html#import-summarise-phyloseq"><i class="fa fa-check"></i><b>7.3</b> Import: Summarise phyloseq</a></li>
<li class="chapter" data-level="7.4" data-path="07-Import.html"><a href="07-Import.html#save-the-phyloseq-object"><i class="fa fa-check"></i><b>7.4</b> Save the phyloseq object</a></li>
<li class="chapter" data-level="7.5" data-path="07-Import.html"><a href="07-Import.html#import-recap"><i class="fa fa-check"></i><b>7.5</b> Import: recap</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html"><i class="fa fa-check"></i><b>8</b> Summarise phyloseq</a>
<ul>
<li class="chapter" data-level="8.1" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#setup"><i class="fa fa-check"></i><b>8.1</b> Setup</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-header"><i class="fa fa-check"></i><b>8.1.1</b> Summarise header</a></li>
<li class="chapter" data-level="8.1.2" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-phyloseq"><i class="fa fa-check"></i><b>8.1.2</b> Summarise phyloseq</a></li>
<li class="chapter" data-level="8.1.3" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#reads-per-sample"><i class="fa fa-check"></i><b>8.1.3</b> Reads per sample</a></li>
<li class="chapter" data-level="8.1.4" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#asvs-per-sample"><i class="fa fa-check"></i><b>8.1.4</b> ASVs per sample</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-recap"><i class="fa fa-check"></i><b>8.2</b> Summarise: recap</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html"><i class="fa fa-check"></i><b>9</b> Minimum read depth</a>
<ul>
<li class="chapter" data-level="9.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#considerations"><i class="fa fa-check"></i><b>9.1</b> Considerations</a></li>
<li class="chapter" data-level="9.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#minimum-read-depth-section-title"><i class="fa fa-check"></i><b>9.2</b> Minimum read depth section title</a></li>
<li class="chapter" data-level="9.3" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#read-depth-vector-and-histogram"><i class="fa fa-check"></i><b>9.3</b> Read depth vector and histogram</a></li>
<li class="chapter" data-level="9.4" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#sample-depth-boxplot"><i class="fa fa-check"></i><b>9.4</b> Sample depth boxplot</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#creating-data-frame-for-boxplots"><i class="fa fa-check"></i><b>9.4.1</b> Creating data frame for boxplots</a></li>
<li class="chapter" data-level="9.4.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#depth-boxplots"><i class="fa fa-check"></i><b>9.4.2</b> Depth boxplots</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#chap9rarefaction"><i class="fa fa-check"></i><b>9.5</b> Rarefaction curve</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#asv-abundance-data-frame"><i class="fa fa-check"></i><b>9.5.1</b> ASV abundance data frame</a></li>
<li class="chapter" data-level="9.5.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#vegans-rarecurve"><i class="fa fa-check"></i><b>9.5.2</b> vegan's rarecurve</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#filtering-by-minimum-read-depth"><i class="fa fa-check"></i><b>9.6</b> Filtering by minimum read depth</a></li>
<li class="chapter" data-level="9.7" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#minmum-read-depth-summary"><i class="fa fa-check"></i><b>9.7</b> Minmum read depth: Summary</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html"><i class="fa fa-check"></i><b>10</b> Relative abundance</a>
<ul>
<li class="chapter" data-level="10.1" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-setup"><i class="fa fa-check"></i><b>10.1</b> Relative abundance: setup</a></li>
<li class="chapter" data-level="10.2" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-transformation"><i class="fa fa-check"></i><b>10.2</b> Relative abundance: transformation</a></li>
<li class="chapter" data-level="10.3" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#rare-asv-removal"><i class="fa fa-check"></i><b>10.3</b> Rare ASV removal</a></li>
<li class="chapter" data-level="10.4" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-number-of-asvs-vector"><i class="fa fa-check"></i><b>10.4</b> Relative abundance: number of ASVs vector</a></li>
<li class="chapter" data-level="10.5" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-save"><i class="fa fa-check"></i><b>10.5</b> Relative abundance: save</a></li>
<li class="chapter" data-level="10.6" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-recap"><i class="fa fa-check"></i><b>10.6</b> Relative abundance: recap</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html"><i class="fa fa-check"></i><b>11</b> Rarefaction</a>
<ul>
<li class="chapter" data-level="11.1" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-setup"><i class="fa fa-check"></i><b>11.1</b> Rarefaction: setup</a></li>
<li class="chapter" data-level="11.2" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve"><i class="fa fa-check"></i><b>11.2</b> Rarefaction curve</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#analogy"><i class="fa fa-check"></i><b>11.2.1</b> Analogy</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve-simple-plot"><i class="fa fa-check"></i><b>11.3</b> Rarefaction curve: simple plot</a></li>
<li class="chapter" data-level="11.4" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve-better-plot"><i class="fa fa-check"></i><b>11.4</b> Rarefaction curve: better plot</a></li>
<li class="chapter" data-level="11.5" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve-display-plot"><i class="fa fa-check"></i><b>11.5</b> Rarefaction curve: display plot</a></li>
<li class="chapter" data-level="11.6" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-slope"><i class="fa fa-check"></i><b>11.6</b> Rarefaction slope</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R community analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="relative-abundance" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Relative abundance<a href="10-Relative_abundance.html#relative-abundance" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this chapter we are going to create a <code>phyloseq</code> with relative abundance values. We would use our read depth filtered data if we wanted to use it for other downstream analyses.
But, as stated in the last chapter all our samples looked good so we will use our original abundance <code>phyloseq</code> object.</p>
<p>Once we have created our relative abundance data we will remove rare ASVs.
We will then check how many ASVs we lost and save the relative abundance <code>phyloseq</code> object.</p>
<div id="relative-abundance-setup" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> Relative abundance: setup<a href="10-Relative_abundance.html#relative-abundance-setup" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The first steps before analysis are:</p>
<ul>
<li>Create a new notebook called <strong>"3-relative_abundance.ipynb"</strong>.</li>
<li>Add a markdown cell with the first level header: <code># Relative abundance</code>.</li>
<li>Add the below to a code cell to load in the <code>phyloseq</code> object and libraries.</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="10-Relative_abundance.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb39-2"><a href="10-Relative_abundance.html#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb39-3"><a href="10-Relative_abundance.html#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb39-4"><a href="10-Relative_abundance.html#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb39-5"><a href="10-Relative_abundance.html#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span></code></pre></div>
<p>From now on you will get less instructions on your notebook structure.
Please create your own coding and markdown cells where you think appropriate.</p>
</div>
<div id="relative-abundance-transformation" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> Relative abundance: transformation<a href="10-Relative_abundance.html#relative-abundance-transformation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that we have the data loaded we can create a new <code>phyloseq</code> object by converting the abundance values to relative abundance.</p>
<p>This is carried out with the <code>microbiome</code> function <code>transform()</code>. With it we transform the ASV abundance table within to a "compositional" table (relative abundance).</p>
<p>Run the below command in an appropriate place in your notebook:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="10-Relative_abundance.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Convert abundance table to relative abundance (compositional) table</span></span>
<span id="cb40-2"><a href="10-Relative_abundance.html#cb40-2" aria-hidden="true" tabindex="-1"></a>pseq_relabund <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">transform</span>(pseq, <span class="st">&quot;compositional&quot;</span>)</span></code></pre></div>
<p>As always it is good to check the contents of the new ASV table.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="10-Relative_abundance.html#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise and check sample counts which should each amount to 1</span></span>
<span id="cb41-2"><a href="10-Relative_abundance.html#cb41-2" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq_relabund)</span>
<span id="cb41-3"><a href="10-Relative_abundance.html#cb41-3" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_relabund)</span></code></pre></div>
<p>You will notice the read count for each sample is <strong>1</strong>. This abundance table contains fractional relative abundances for each ASV. This fraction is relative to the total abundance within each sample. Therefore, all the fractional relative abundance of ASVs in one sample total 1.</p>
</div>
<div id="rare-asv-removal" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> Rare ASV removal<a href="10-Relative_abundance.html#rare-asv-removal" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>When we are using total or rarefied abundance values we want to keep rare ASVs, singletons, and doubletons. These are important to keep so we can get detailed data. Additionally, some alpha diversity metrics require these values.</p>
<div class="webex-solution">
<button>
Singletons &amp; Doubletons defintion
</button>
<ul>
<li><strong>Singleton</strong>: A single, unique DNA sequence. I.e. an ASV with a total abundance of 1 across all samples.</li>
<li><strong>Doubleton</strong>: A DNA sequence that only appears twice. I.e. an ASV with a total abundance of 2 across all samples.</li>
</ul>
</div>
<p>However, when we are using relative abundance data it can be a good idea to remove rare features (ASVs). Relative abundance data is good to get a broader picture and therefore rare ASVs may add too much noise.</p>
<p>We will therefore remove ASVs with low relative abundances. A recommended method is to remove ASVs with a mean relative abundance equal to or less than <code>1e-5</code> (0.00001).</p>
<p>For this we will use <code>phyloseq::filter_taxa()</code>. This uses a function to keep ASVs with a mean relative abundance &gt; 1e-5.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="10-Relative_abundance.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove rare ASVs, retain ASVs with mean &gt; 1e-5</span></span>
<span id="cb42-2"><a href="10-Relative_abundance.html#cb42-2" aria-hidden="true" tabindex="-1"></a>pseq_relabund <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">filter_taxa</span>(</span>
<span id="cb42-3"><a href="10-Relative_abundance.html#cb42-3" aria-hidden="true" tabindex="-1"></a>  pseq_relabund, <span class="cf">function</span>(x) <span class="fu">mean</span>(x) <span class="sc">&gt;</span> <span class="fl">1e-5</span>, <span class="cn">TRUE</span></span>
<span id="cb42-4"><a href="10-Relative_abundance.html#cb42-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>It is very important to make sure only a small amount of our relative abundance has been lost across our samples. Check this with the following commands.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="10-Relative_abundance.html#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Summarise and check sample counts which should each amount to around 1</span></span>
<span id="cb43-2"><a href="10-Relative_abundance.html#cb43-2" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">summarize_phyloseq</span>(pseq_relabund)</span>
<span id="cb43-3"><a href="10-Relative_abundance.html#cb43-3" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq_relabund)</span></code></pre></div>
<ul>
<li>What is the approximate minimum total relative abundance?
<div id="radio_WSNQWHEKBQ" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_WSNQWHEKBQ" value="answer"></input> <span><strong>0.979</strong></span></label><label><input type="radio" autocomplete="off" name="radio_WSNQWHEKBQ" value=""></input> <span><strong>0.997</strong></span></label><label><input type="radio" autocomplete="off" name="radio_WSNQWHEKBQ" value=""></input> <span><strong>1</strong></span></label>
</div></li>
<li>What is the approximate average total relative abundance?
<div id="radio_XEMOTBUXRS" class="webex-radiogroup">
<label><input type="radio" autocomplete="off" name="radio_XEMOTBUXRS" value=""></input> <span><strong>0.978</strong></span></label><label><input type="radio" autocomplete="off" name="radio_XEMOTBUXRS" value="answer"></input> <span><strong>0.997</strong></span></label><label><input type="radio" autocomplete="off" name="radio_XEMOTBUXRS" value=""></input> <span><strong>1</strong></span></label>
</div></li>
</ul>
<p>We have lost a very small amount of relative abundance with no sample having lost more than 0.04 (4%). Most of the samples have lost less than 0.01 (1%). For the purposes of our relatve abundance data this is fine.</p>
<p>If you were worried you had lost too much you could try to be less stringent and use a smaller value such as <code>1e-6</code>. If that is still to strict try <code>1e-7</code> etc.. If you do this ensure you load the abundance <code>phyloseq</code> object again so you are not filtering the already filtered data.</p>
</div>
<div id="relative-abundance-number-of-asvs-vector" class="section level2 hasAnchor" number="10.4">
<h2><span class="header-section-number">10.4</span> Relative abundance: number of ASVs vector<a href="10-Relative_abundance.html#relative-abundance-number-of-asvs-vector" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>It is useful to keep track of how many ASVs are in our various <code>phyloseq</code> objects. Use the below commands to load in the vector we created in chapter 8. This contains the number of ASVs in our original data. We'll add the number of ASVs we have in our relative abundance table.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="10-Relative_abundance.html#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Load in ASV count vector</span></span>
<span id="cb44-2"><a href="10-Relative_abundance.html#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;num_asvs_vec.RData&quot;</span>)</span>
<span id="cb44-3"><a href="10-Relative_abundance.html#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Add relative abundance ASV count</span></span>
<span id="cb44-4"><a href="10-Relative_abundance.html#cb44-4" aria-hidden="true" tabindex="-1"></a>num_asvs_vec[<span class="st">&quot;relabund&quot;</span>] <span class="ot">&lt;-</span> <span class="fu">nrow</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq_relabund))</span>
<span id="cb44-5"><a href="10-Relative_abundance.html#cb44-5" aria-hidden="true" tabindex="-1"></a>num_asvs_vec</span>
<span id="cb44-6"><a href="10-Relative_abundance.html#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Check how many ASVs lost</span></span>
<span id="cb44-7"><a href="10-Relative_abundance.html#cb44-7" aria-hidden="true" tabindex="-1"></a>num_asvs_vec[<span class="st">&quot;abundance&quot;</span>] <span class="sc">-</span> num_asvs_vec[<span class="st">&quot;relabund&quot;</span>]</span>
<span id="cb44-8"><a href="10-Relative_abundance.html#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Save vector as a new file</span></span>
<span id="cb44-9"><a href="10-Relative_abundance.html#cb44-9" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(num_asvs_vec, <span class="at">file=</span><span class="st">&quot;num_asvs_vec.v2.RData&quot;</span>)</span>
<span id="cb44-10"><a href="10-Relative_abundance.html#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Remove object from environment</span></span>
<span id="cb44-11"><a href="10-Relative_abundance.html#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(num_asvs_vec)</span></code></pre></div>
<p>We saved the vector as a new file. This is in case we need to rerun the code in <strong>"2-preprocess.ipynb"</strong>. If we did that and the files had the same name we could overwrite the file and it may be confusing to know what happened to it.</p>
</div>
<div id="relative-abundance-save" class="section level2 hasAnchor" number="10.5">
<h2><span class="header-section-number">10.5</span> Relative abundance: save<a href="10-Relative_abundance.html#relative-abundance-save" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We have created our relative abundance table and we are happy with it. Therefore, we will save the <code>phyloseq</code> object.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="10-Relative_abundance.html#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Save relative abundance phyloseq</span></span>
<span id="cb45-2"><a href="10-Relative_abundance.html#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(pseq_relabund, <span class="at">file =</span> <span class="st">&quot;phyloseq_relabund.RData&quot;</span>)</span></code></pre></div>
<p>Once you have donw this you can save, then close and halt the notebook.</p>
</div>
<div id="relative-abundance-recap" class="section level2 hasAnchor" number="10.6">
<h2><span class="header-section-number">10.6</span> Relative abundance: recap<a href="10-Relative_abundance.html#relative-abundance-recap" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We have created and saved our relative abundance <code>phyloseq</code> object for later use. IN addition, we removed rare ASVs since we will use this data to look at broad strokes such as taxonomc tables.</p>

</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  if (t = document.getElementById("webex-total_correct")) {
    var correct = document.getElementsByClassName("webex-correct").length;
    var solvemes = document.getElementsByClassName("webex-solveme").length;
    var radiogroups = document.getElementsByClassName("webex-radiogroup").length;
    var selects = document.getElementsByClassName("webex-select").length;
    
    t.innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");
  
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");
  
  var cl = this.classList
  
  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;
  
  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }
  
  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

window.onload = function() {
  console.log("onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }
  
  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="09-Minimum_read_depth.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="11-Rarefaction.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
