<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 11 Rarefaction | R community analysis</title>
  <meta name="description" content="NEOF book for the R community analysis workflow" />
  <meta name="generator" content="bookdown 0.31 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 11 Rarefaction | R community analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the R community analysis workflow" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 11 Rarefaction | R community analysis" />
  
  <meta name="twitter:description" content="NEOF book for the R community analysis workflow" />
  <meta name="twitter:image" content="/figures/NEOF.png" />

<meta name="author" content="Matthew R. Gemmell" />


<meta name="date" content="2023-03-13" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="10-Relative_abundance.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/neof_small_logo.png"></a></li>
<li><a>R community analysis</a></li>

<li class="divider"></li>
<li class="part"><span><b>Intro</b></span></li>
<li class="chapter" data-level="1" data-path="01-R_community_main_workflow.html"><a href="01-R_community_main_workflow.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html"><i class="fa fa-check"></i><b>2</b> Dataset &amp; workflow</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#dataset"><i class="fa fa-check"></i><b>2.1</b> Dataset</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#sites"><i class="fa fa-check"></i><b>2.1.1</b> Sites</a></li>
<li class="chapter" data-level="2.1.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#culture-media"><i class="fa fa-check"></i><b>2.1.2</b> Culture media</a></li>
<li class="chapter" data-level="2.1.3" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#summary-questions"><i class="fa fa-check"></i><b>2.1.3</b> Summary &amp; questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="02-Dataset_and_workflow.html"><a href="02-Dataset_and_workflow.html#workflow"><i class="fa fa-check"></i><b>2.2</b> Workflow</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03-R_packages.html"><a href="03-R_packages.html"><i class="fa fa-check"></i><b>3</b> R Packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03-R_packages.html"><a href="03-R_packages.html#r-packageslibraries"><i class="fa fa-check"></i><b>3.1</b> R packages/libraries</a></li>
<li class="chapter" data-level="3.2" data-path="03-R_packages.html"><a href="03-R_packages.html#the-grammar-of-graphics"><i class="fa fa-check"></i><b>3.2</b> The grammar of graphics</a></li>
<li class="chapter" data-level="3.3" data-path="03-R_packages.html"><a href="03-R_packages.html#phyloseq"><i class="fa fa-check"></i><b>3.3</b> phyloseq</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Setup.html"><a href="04-Setup.html"><i class="fa fa-check"></i><b>4</b> Set-up</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Setup.html"><a href="04-Setup.html#cluster"><i class="fa fa-check"></i><b>4.1</b> Logon instructions</a></li>
<li class="chapter" data-level="4.2" data-path="04-Setup.html"><a href="04-Setup.html#mamba"><i class="fa fa-check"></i><b>4.2</b> Mamba</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="05-Jupyter.html"><a href="05-Jupyter.html"><i class="fa fa-check"></i><b>5</b> Jupyter</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-Jupyter.html"><a href="05-Jupyter.html#open-jupyter-notebook"><i class="fa fa-check"></i><b>5.1</b> Open Jupyter-notebook</a></li>
<li class="chapter" data-level="5.2" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-r-notebook"><i class="fa fa-check"></i><b>5.2</b> Create R notebook</a></li>
<li class="chapter" data-level="5.3" data-path="05-Jupyter.html"><a href="05-Jupyter.html#cells-and-code"><i class="fa fa-check"></i><b>5.3</b> Cells and code</a></li>
<li class="chapter" data-level="5.4" data-path="05-Jupyter.html"><a href="05-Jupyter.html#create-new-cells"><i class="fa fa-check"></i><b>5.4</b> Create new cells</a></li>
<li class="chapter" data-level="5.5" data-path="05-Jupyter.html"><a href="05-Jupyter.html#running-code"><i class="fa fa-check"></i><b>5.5</b> Running code</a></li>
<li class="chapter" data-level="5.6" data-path="05-Jupyter.html"><a href="05-Jupyter.html#saving-the-file"><i class="fa fa-check"></i><b>5.6</b> Saving the file</a></li>
<li class="chapter" data-level="5.7" data-path="05-Jupyter.html"><a href="05-Jupyter.html#title-cells-with-markdown"><i class="fa fa-check"></i><b>5.7</b> Title cells with markdown</a></li>
<li class="chapter" data-level="5.8" data-path="05-Jupyter.html"><a href="05-Jupyter.html#close-the-notebook"><i class="fa fa-check"></i><b>5.8</b> Close the notebook</a></li>
<li class="chapter" data-level="5.9" data-path="05-Jupyter.html"><a href="05-Jupyter.html#video-tutorial"><i class="fa fa-check"></i><b>5.9</b> Video tutorial</a></li>
</ul></li>
<li class="part"><span><b>Data preparation</b></span></li>
<li class="chapter" data-level="6" data-path="06-Preprocess_part.html"><a href="06-Preprocess_part.html"><i class="fa fa-check"></i><b>6</b> Data prep intro</a></li>
<li class="chapter" data-level="7" data-path="07-Import.html"><a href="07-Import.html"><i class="fa fa-check"></i><b>7</b> Import</a>
<ul>
<li class="chapter" data-level="7.1" data-path="07-Import.html"><a href="07-Import.html#import-notebook"><i class="fa fa-check"></i><b>7.1</b> Import: notebook</a></li>
<li class="chapter" data-level="7.2" data-path="07-Import.html"><a href="07-Import.html#qiime2r"><i class="fa fa-check"></i><b>7.2</b> qiime2R</a></li>
<li class="chapter" data-level="7.3" data-path="07-Import.html"><a href="07-Import.html#import-summarise-phyloseq"><i class="fa fa-check"></i><b>7.3</b> Import: Summarise phyloseq</a></li>
<li class="chapter" data-level="7.4" data-path="07-Import.html"><a href="07-Import.html#save-the-phyloseq-object"><i class="fa fa-check"></i><b>7.4</b> Save the phyloseq object</a></li>
<li class="chapter" data-level="7.5" data-path="07-Import.html"><a href="07-Import.html#import-recap"><i class="fa fa-check"></i><b>7.5</b> Import: recap</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html"><i class="fa fa-check"></i><b>8</b> Summarise phyloseq</a>
<ul>
<li class="chapter" data-level="8.1" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#setup"><i class="fa fa-check"></i><b>8.1</b> Setup</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-header"><i class="fa fa-check"></i><b>8.1.1</b> Summarise header</a></li>
<li class="chapter" data-level="8.1.2" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-phyloseq"><i class="fa fa-check"></i><b>8.1.2</b> Summarise phyloseq</a></li>
<li class="chapter" data-level="8.1.3" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#reads-per-sample"><i class="fa fa-check"></i><b>8.1.3</b> Reads per sample</a></li>
<li class="chapter" data-level="8.1.4" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#asvs-per-sample"><i class="fa fa-check"></i><b>8.1.4</b> ASVs per sample</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="08-Summarise_phyloseq.html"><a href="08-Summarise_phyloseq.html#summarise-recap"><i class="fa fa-check"></i><b>8.2</b> Summarise: recap</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html"><i class="fa fa-check"></i><b>9</b> Minimum read depth</a>
<ul>
<li class="chapter" data-level="9.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#considerations"><i class="fa fa-check"></i><b>9.1</b> Considerations</a></li>
<li class="chapter" data-level="9.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#minimum-read-depth-section-title"><i class="fa fa-check"></i><b>9.2</b> Minimum read depth section title</a></li>
<li class="chapter" data-level="9.3" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#read-depth-vector-and-histogram"><i class="fa fa-check"></i><b>9.3</b> Read depth vector and histogram</a></li>
<li class="chapter" data-level="9.4" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#sample-depth-boxplot"><i class="fa fa-check"></i><b>9.4</b> Sample depth boxplot</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#creating-data-frame-for-boxplots"><i class="fa fa-check"></i><b>9.4.1</b> Creating data frame for boxplots</a></li>
<li class="chapter" data-level="9.4.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#depth-boxplots"><i class="fa fa-check"></i><b>9.4.2</b> Depth boxplots</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#chap9rarefaction"><i class="fa fa-check"></i><b>9.5</b> Rarefaction curve</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#asv-abundance-data-frame"><i class="fa fa-check"></i><b>9.5.1</b> ASV abundance data frame</a></li>
<li class="chapter" data-level="9.5.2" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#vegans-rarecurve"><i class="fa fa-check"></i><b>9.5.2</b> vegan's rarecurve</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#filtering-by-minimum-read-depth"><i class="fa fa-check"></i><b>9.6</b> Filtering by minimum read depth</a></li>
<li class="chapter" data-level="9.7" data-path="09-Minimum_read_depth.html"><a href="09-Minimum_read_depth.html#minmum-read-depth-summary"><i class="fa fa-check"></i><b>9.7</b> Minmum read depth: Summary</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html"><i class="fa fa-check"></i><b>10</b> Relative abundance</a>
<ul>
<li class="chapter" data-level="10.1" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-setup"><i class="fa fa-check"></i><b>10.1</b> Relative abundance: setup</a></li>
<li class="chapter" data-level="10.2" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-transformation"><i class="fa fa-check"></i><b>10.2</b> Relative abundance: transformation</a></li>
<li class="chapter" data-level="10.3" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#rare-asv-removal"><i class="fa fa-check"></i><b>10.3</b> Rare ASV removal</a></li>
<li class="chapter" data-level="10.4" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-number-of-asvs-vector"><i class="fa fa-check"></i><b>10.4</b> Relative abundance: number of ASVs vector</a></li>
<li class="chapter" data-level="10.5" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-save"><i class="fa fa-check"></i><b>10.5</b> Relative abundance: save</a></li>
<li class="chapter" data-level="10.6" data-path="10-Relative_abundance.html"><a href="10-Relative_abundance.html#relative-abundance-recap"><i class="fa fa-check"></i><b>10.6</b> Relative abundance: recap</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html"><i class="fa fa-check"></i><b>11</b> Rarefaction</a>
<ul>
<li class="chapter" data-level="11.1" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-setup"><i class="fa fa-check"></i><b>11.1</b> Rarefaction: setup</a></li>
<li class="chapter" data-level="11.2" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve"><i class="fa fa-check"></i><b>11.2</b> Rarefaction curve</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#analogy"><i class="fa fa-check"></i><b>11.2.1</b> Analogy</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve-simple-plot"><i class="fa fa-check"></i><b>11.3</b> Rarefaction curve: simple plot</a></li>
<li class="chapter" data-level="11.4" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve-better-plot"><i class="fa fa-check"></i><b>11.4</b> Rarefaction curve: better plot</a></li>
<li class="chapter" data-level="11.5" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-curve-display-plot"><i class="fa fa-check"></i><b>11.5</b> Rarefaction curve: display plot</a></li>
<li class="chapter" data-level="11.6" data-path="11-Rarefaction.html"><a href="11-Rarefaction.html#rarefaction-slope"><i class="fa fa-check"></i><b>11.6</b> Rarefaction slope</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R community analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rarefaction_chap" class="section level1 hasAnchor" number="11">
<h1><span class="header-section-number">Chapter 11</span> Rarefaction<a href="11-Rarefaction.html#rarefaction_chap" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>We are going to create a rarefied abundance table in this chapter. This
can be a controversial topic so please make your own decision if you
want to do this or not in your own analysis.</p>
<div id="rarefaction-setup" class="section level2 hasAnchor" number="11.1">
<h2><span class="header-section-number">11.1</span> Rarefaction: setup<a href="11-Rarefaction.html#rarefaction-setup" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We will use a new notebook called <strong>"4-rarefaction.ipynb"</strong>. Add the
following to this new notebook to load the required libraries/packages
and data.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="11-Rarefaction.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Libraries</span></span>
<span id="cb46-2"><a href="11-Rarefaction.html#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb46-3"><a href="11-Rarefaction.html#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;microbiome&quot;</span>)</span>
<span id="cb46-4"><a href="11-Rarefaction.html#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;vegan&quot;</span>)</span>
<span id="cb46-5"><a href="11-Rarefaction.html#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;IRdisplay&quot;</span>)</span>
<span id="cb46-6"><a href="11-Rarefaction.html#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Load phyloseq object</span></span>
<span id="cb46-7"><a href="11-Rarefaction.html#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;phyloseq.RData&quot;</span>)</span>
<span id="cb46-8"><a href="11-Rarefaction.html#cb46-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Load ASV count vector</span></span>
<span id="cb46-9"><a href="11-Rarefaction.html#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;num_asvs_vec.v2.RData&quot;</span>)</span></code></pre></div>
<p>Ensure you add markdown and code cells to this notebook to give yourself
a good structure.</p>
</div>
<div id="rarefaction-curve" class="section level2 hasAnchor" number="11.2">
<h2><span class="header-section-number">11.2</span> Rarefaction curve<a href="11-Rarefaction.html#rarefaction-curve" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/rarefaction.png" width="15%" style="display: block; margin: auto;" /></p>
<p>Prior to rarefying our data we need to determine the depth we want to
use. We can carry this out with a rarefaction curve.</p>
<p>A rarefaction curve is produced by randomly sampling the sequences in a
sample (without replacement). The rarefaction curve extracts and plots
the first N sequences of each sample. N is equal to the step size which
we will set as 50 (<code>step = 50</code>).</p>
<p>From these sequences it will plot the number of total ASVs it has found
against the depth it has currently sampled. Therefore, after 10 steps it
will be at a depth of 500 on the x-axis (50 * 10) and it will show how
many unique ASVs it has currently discovered from all 10 samplings.</p>
<div id="analogy" class="section level3 hasAnchor" number="11.2.1">
<h3><span class="header-section-number">11.2.1</span> Analogy<a href="11-Rarefaction.html#analogy" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You are cataloguing the bird diversity (ASV diversity) in three forests/woods
in Hampshire (3 samples of our abundance table). You record your
findings over a week for each forest. Your record on the cumulative
amount of unique species you find over the 3 separate weeks are in the
table below:</p>
<table>
<caption>Cumulative bird species found in Hampshire forests/woods over a week</caption>
<colgroup>
<col width="20%" />
<col width="24%" />
<col width="27%" />
<col width="27%" />
</colgroup>
<thead>
<tr class="header">
<th>Day</th>
<th>New Forest (93900 acres)</th>
<th>Telegraph woods (50 acres)</th>
<th>Old Lords wood (4.67 acres)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>12</td>
<td>7</td>
<td>8</td>
</tr>
<tr class="even">
<td>2</td>
<td>21</td>
<td>12</td>
<td>10</td>
</tr>
<tr class="odd">
<td>3</td>
<td>32</td>
<td>16</td>
<td>11</td>
</tr>
<tr class="even">
<td>4</td>
<td>40</td>
<td>21</td>
<td>13</td>
</tr>
<tr class="odd">
<td>5</td>
<td>49</td>
<td>23</td>
<td>13</td>
</tr>
<tr class="even">
<td>6</td>
<td>60</td>
<td>24</td>
<td>13</td>
</tr>
<tr class="odd">
<td>7</td>
<td>65</td>
<td>25</td>
<td>13</td>
</tr>
</tbody>
</table>
<p><img src="R-community-analysis_files/figure-html/unnamed-chunk-116-1.png" width="672" /></p>
<p>In this case we can be fairly confident that seven days is enough to find all the birds we can in the Old Lords wood.
I.e. the sampling depth is enough.
Just like bird watching, it is very hard to find all the species present in a sample with metabarcoding or shotgun metagenomics.
This can be caused by our barcodes not picking up all species, our sampling wasn't perfect, some species are very rare, or errors that DNA sequencing introduces.</p>
<p>A week seems like a good amount for sampling the Telegraph woods but it is possible that some more species have not been recorded.</p>
<p>Strikingly, a week does not appear to be enough time for the New Forest. The amount of new species decreased to 5 on day 7 but the curve has not plateau'd.</p>
<p>This hopefully helps you understand why rarefaction curves are a useful measure of how well we have captured the diversity of samples at differnet depths.</p>
</div>
</div>
<div id="rarefaction-curve-simple-plot" class="section level2 hasAnchor" number="11.3">
<h2><span class="header-section-number">11.3</span> Rarefaction curve: simple plot<a href="11-Rarefaction.html#rarefaction-curve-simple-plot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We'll use <code>vegan</code> to create our rarefaction curve like we did in
<a href="09-Minimum_read_depth.html#chap9rarefaction">Chapter 9</a>.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="11-Rarefaction.html#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract ASV table as data frame</span></span>
<span id="cb47-2"><a href="11-Rarefaction.html#cb47-2" aria-hidden="true" tabindex="-1"></a>asv_abund_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(pseq)))</span>
<span id="cb47-3"><a href="11-Rarefaction.html#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Rarefaction curve</span></span>
<span id="cb47-4"><a href="11-Rarefaction.html#cb47-4" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">rarecurve</span>(</span>
<span id="cb47-5"><a href="11-Rarefaction.html#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> asv_abund_df, <span class="at">step =</span> <span class="dv">50</span>,</span>
<span id="cb47-6"><a href="11-Rarefaction.html#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Read depth&quot;</span>,</span>
<span id="cb47-7"><a href="11-Rarefaction.html#cb47-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;ASVs&quot;</span></span>
<span id="cb47-8"><a href="11-Rarefaction.html#cb47-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="rarefaction-curve-better-plot" class="section level2 hasAnchor" number="11.4">
<h2><span class="header-section-number">11.4</span> Rarefaction curve: better plot<a href="11-Rarefaction.html#rarefaction-curve-better-plot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>It is a useful plot but we can make it bigger and better! We will do
that with the following additions:</p>
<ul>
<li>Saving it as a <code>.png</code> file with the functions <code>png()</code> and
<code>dev.off()</code>.</li>
<li>Adding extra options to <code>vegan::rarecurve()</code>:
<ul>
<li><code>lwd</code>: Sets the line width of the plot.</li>
<li><code>label</code>: Turn the sample labels on (<code>T</code>) or off (<code>F</code>).</li>
<li><code>sample</code>: Draws a vertical line at the specified depth.
Additionally, draws a horizontal line for each sample showing
how many ASVs were discovered at the sampling depth.</li>
</ul></li>
<li>Adding a vertical line showing the minimum depth of the samples with
<code>abline()</code>.</li>
</ul>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="11-Rarefaction.html#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Improved plot saved as file</span></span>
<span id="cb48-2"><a href="11-Rarefaction.html#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">filename =</span> <span class="st">&quot;./rarefaction_plot.png&quot;</span>, <span class="at">res =</span> <span class="dv">300</span>,</span>
<span id="cb48-3"><a href="11-Rarefaction.html#cb48-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">units =</span> <span class="st">&quot;mm&quot;</span>, <span class="at">height =</span> <span class="dv">200</span>, <span class="at">width =</span> <span class="dv">300</span>)</span>
<span id="cb48-4"><a href="11-Rarefaction.html#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot</span></span>
<span id="cb48-5"><a href="11-Rarefaction.html#cb48-5" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">rarecurve</span>(</span>
<span id="cb48-6"><a href="11-Rarefaction.html#cb48-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> asv_abund_df, <span class="at">step =</span> <span class="dv">50</span>,</span>
<span id="cb48-7"><a href="11-Rarefaction.html#cb48-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;Read depth&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;ASVs&quot;</span>, <span class="at">lwd=</span><span class="dv">1</span>, <span class="at">label =</span> F,</span>
<span id="cb48-8"><a href="11-Rarefaction.html#cb48-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample =</span> <span class="fu">min</span>(microbiome<span class="sc">::</span><span class="fu">readcount</span>(pseq))</span>
<span id="cb48-9"><a href="11-Rarefaction.html#cb48-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb48-10"><a href="11-Rarefaction.html#cb48-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div id="rarefaction-curve-display-plot" class="section level2 hasAnchor" number="11.5">
<h2><span class="header-section-number">11.5</span> Rarefaction curve: display plot<a href="11-Rarefaction.html#rarefaction-curve-display-plot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We have created and saved the plot as a <code>.png</code> but it is not being
displayed. We can use the package
<a href="https://github.com/IRkernel/IRdisplay"><code>IRdisplay</code></a> to display plots
from files in <code>jupyter-notebook</code>. In this case the function
<code>display_png()</code> is used.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="11-Rarefaction.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Display the plot in jupyter notebook</span></span>
<span id="cb49-2"><a href="11-Rarefaction.html#cb49-2" aria-hidden="true" tabindex="-1"></a>IRdisplay<span class="sc">::</span><span class="fu">display_png</span>(<span class="at">file=</span><span class="st">&quot;./rarefaction_plot.png&quot;</span>)</span></code></pre></div>
</div>
<div id="rarefaction-slope" class="section level2 hasAnchor" number="11.6">
<h2><span class="header-section-number">11.6</span> Rarefaction slope<a href="11-Rarefaction.html#rarefaction-slope" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>#Rarefy abundance table #### #i.e. convert abundance numbers so each
sample has equal depth</p>
<p>#Before rarefying a table it is good to make a rarefaction curve #This
is to help us choose an appropriate rarefaction threshold</p>
<p>#We will use the very useful package vegan #Ignore any warning message
vegan::rarecurve( x = as.data.frame(t(phyloseq::otu_table(pseq))), step
= 50)</p>
<p>#Let us improve this and save it into a file png(filename =
"./rarefaction_plot.png", res = 300, units = "mm", height = 200, width =
300) vegan::rarecurve( x = as.data.frame(t(phyloseq::otu_table(pseq))),
step = 50, ylab="ASVs", lwd=1,label=F) #Add a vertical line of the
smallest sample depth abline(v = min(reads_sample), col="red") dev.off()</p>
<p>#With this we can see that the majorty of samples plateau at #the
minimum sampleing depth #Therefore we can use this as a rarefaction size
pseq_rarefy &lt;- phyloseq::rarefy_even_depth( pseq, sample.size =
min(reads_sample), rngseed = 1000)</p>
<p>#Summarise and check sample counts which should each amount to 10433
microbiome::summarize_phyloseq(pseq_rarefy)
microbiome::readcount(pseq_rarefy)</p>
<p>#Check ASVs num_asvs_vec["rarefied"] &lt;-
nrow(phyloseq::otu_table(pseq_rarefy)) num_asvs_vec</p>
<p>#Save phyloseq object save(pseq_relpseq_rarefyabund, file =
"phyloseq_rarefied.RData")</p>

</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  if (t = document.getElementById("webex-total_correct")) {
    var correct = document.getElementsByClassName("webex-correct").length;
    var solvemes = document.getElementsByClassName("webex-solveme").length;
    var radiogroups = document.getElementsByClassName("webex-radiogroup").length;
    var selects = document.getElementsByClassName("webex-select").length;
    
    t.innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");
  
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");
  
  var cl = this.classList
  
  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;
  
  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }
  
  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }
  
  update_total_correct();
}

window.onload = function() {
  console.log("onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }
  
  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="10-Relative_abundance.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
