# Importing {#importchap}
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/data.png", auto_pdf = TRUE)
``` 

Apart from `QIIME2` artifact files, other file types from other programs cann be imported into a `phyloseq` object.

## Bracken
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/bracken.png", auto_pdf = TRUE)
``` 

A common way to carry out shotgun metagenomics taxonomic abundance profiling is with `Kraken2` and `Bracken`.
This section will show you how to import the `bracken` output.

### Bracken: setup
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/setup_3.png", auto_pdf = TRUE)
``` 

First we need the `Bracken` output in the `Kraken2` report output file format.
We can copy over the output made during the NEOF Shotgun metagenomics workshop.

__Note__: If you are using `Bracken` ensure you are using the latest version as older versions do not provide this output.

```{bash, eval=FALSE}
mkdir bracken_r_comm
cd bracken_r_comm
cp /pub14/tea/nsc206/NEOF/R_community/bracken_reports/*breport2 .
```

We can also copy over the metadata file which we will use later.

```{bash, eval=FALSE}
cp /pub14/tea/nsc206/NEOF/R_community/bracken_reports/KW_metadata.tsv .
```

### Bracken: Convert to biom
```{r, fig.align = 'center',out.width= '8%', echo=FALSE }
knitr::include_graphics(path = "figures/biom_tie.png", auto_pdf = TRUE)
```

Unfortunately we can't use the `bracken` report files.
Instead we need to convert them to a single collated biom file with `kraken-biom`.

```{bash, eval=FALSE}
kraken-biom *breport2 -o KW.biom 
```

[__Biom__](https://biom-format.org/) files are not human readable.
Therefore to check the file is correct we can create a tsv (tab seperated value) file to inspect.

```{bash, eval=FALSE}
#Create tsv file
biom convert -i KW.biom -o KW.tsv --header-key taxonomy--to-tsv
#inspect tsv file
less KW.tsv
```

If your file has a header starting with "#OTU ID" and the sample names, followed by rows of OTUs with abundance numbers it should be good.
You'll also have the taxonomy lineage as the last column because we added in `--header-key taxonomy`.

### Jupyter notebook
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/jupyter_notebook.png", auto_pdf = TRUE)
```

Now we are ready to carry out some R analysis.

Open up `jupyter-notebook` and create a new `R` notebook called "Bracken_phyloseq_creation".

At the top of the notebook create a cell to load the libaries we'll need.

```{bash, eval=FALSE}
#Libraries
library("phyloseq")
library("microbiome")
```

#### Import biom into phyloseq {-}
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/phyloseq_logo.png", auto_pdf = TRUE)
```

The first step is to import the biom file.

```{r, eval=FALSE}
#Read in bracken biom file
pseq <- phyloseq::import_biom("/KW.biom")
```

#### Import sample data {-}
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/samples.png", auto_pdf = TRUE)
```

Next we need to import the samples data.
We have a tab delimited file we will read into R as a __data frame__.
The we can assign the sample data of `pseq` to the __data frame__.
As long as the row names contain the same sample names as in the biom file it should work correctly.

```{R, eval=FALSE}
#Read in metadata
metadf <- read.csv("KW_metadata.tsv", sep="\t", check.names = FALSE, row.names=1)
#Check metadf
head(metadf)
#Assign sample data
phyloseq::sample_data(pseq <- metadf)
```

We can then summarise our `phyloseq` object to make sure it seems ok.

```{R, eval=FALSE}
#Summarize
microbiome::summarize_phyloseq(pseq)
```

#### Taxonomy {-}
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/taxa.png", auto_pdf = TRUE)
```

Let's check out the taxonomy

```{R, eval=FALSE}
#Check taxonomy
head(phyloseq::tax_table(pseq))
```

It looks good but we'll carry out 2 tasks.

First we'll change the taxonomy headers from Rank1... to Kingdom....
We know we can do this sue to the prefix of the taxonomies ("k__", "p__"...)

```{r, eval=FALSE}
#Change tax column names
colnames(phyloseq::tax_table(pseq)) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
head(phyloseq::tax_table(pseq))
```

This next step is optional.
We are going to aggregate the `phyloseq` object from taxids to genus.
Shotgun metagenomics is not always the best tool at differentiating between species so genus classifications will be more reliable.

```{r, eval=FALSE}
#Aggregate to species
genus_pseq <- microbiome::aggregate_taxa(pseq, "Genus", verbose = FALSE)
#Summarise
microbiome::summarize_phyloseq(genus_pseq)
#Check number of genera
paste0("Genera: ", nrwo(phyloseq::otu_table(genus_pseq)))
```

In your own analysis you may even want to use family or order to aggregate to, or to keep using species.
Know your data, look at the literature, and make an informed decision.
Of course it can be useful to look at multiple levels of taxonomy.

## Plain text files
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/plain_text.png", auto_pdf = TRUE)
```

If you have plain text files you can read them into R and then phyloseq-ize them.
The phyloseq website has a nice tutorial on how to carry this out: https://joey711.github.io/phyloseq/import-data.


