# 16S R packages {#rpackages}
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/R.png", auto_pdf = TRUE)
```  

There are various R packages that can be used to analyse 16S data. Generally these will take the output from QIIME2 so you can carry out further downstream analysis on the QIIME2 created ASV table, taxonomy, and phylogenetic tree.

I would recommend only those with a good foundation in R to use these packages. If you do not have this I would recommend trying to gain it.

## qiime2R
```{r, fig.align = 'center',out.width= '40%', echo=FALSE }
knitr::include_graphics(path = "figures/qiime2r.png", auto_pdf = TRUE)
```  

https://github.com/jbisanz/qiime2R

qiime2R is an R package to import QIIME2 artifacts into R compatible files. It can convert artifacts into data structures that are useable in R. Additionally it can also create some plots.

One of its most useful commands is `qza_to_phyloseq()` which converts a .qza artifact into the phyloseq data structure.

## phyloseq
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/phyloseq_logo.png", auto_pdf = TRUE)
```  

https://joey711.github.io/phyloseq/

phyloseq is a very powerful R package to carry out microbiome profile analysis. It is a great way to produce custom plots for alpha and beta diversity, heatmaps, network plotting, etc. It can be used to analyse any type of community based data such as shotgun metagenomic data.

There are two large caveats:

1. It can be quite difficult and laborious to convert data to the required phyloseq R object (a data structure created for phyloseq). qiime2R does make this much easier for 16S data though.
2. The documentation can be quite hard to follow and assumes you already have a lot of specific knowledge. It uses ggplot2 coding conventions so it is essential you learn the R package "ggplot2" prior to trying phyloseq.

## microbiome

https://microbiome.github.io/tutorials/

Once you start using phyloseq you will probably notice there are many commands missing you may require. Thankfully the microbiome R package is a very good supplemental package to phyloseq.

This package supports the phyloseq data format and acts to expand the available toolkit to analyse this data format. I would highly recommend using it if you start using phyloseq.