# Exporting Artifacts {#export}
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/artefact_to_tsv.png", auto_pdf = TRUE)
```  

Artifact files are very useful when carrying out analysis in QIIME2. However they are impossible to use (I think) with anything outside of QIIME2. Therefore it can be very useful to know how to export data out of Artifact files.

This section will show how to export various different artifact files into text files. However, for some data types they must be converted into a BIOM file first. 

## BIOM files
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/biom.png", auto_pdf = TRUE)
```
BIOM stands for BIological Observation Matrix. BIOM files generally contain abundance tables. These files are not human-readable but can be used with certain programs. Here we will only manipulate them in such ways to get the text files we need.

For more information on the BIOM format please see: https://biom-format.org/

We will be using the command `biom convert` to convert BIOM files to TSV (Tab Separated Value) files. For more info on this please see: https://biom-format.org/documentation/biom_conversion.html

## Setup
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/setup.png", auto_pdf = TRUE)
```
First copy a new directory that will contain all the QIIME2 output you need and then change directory to it.

```{bash eval=FALSE}
cp -r /pub39/tea/matthew/NEOF/16s_workshop/16s_export ~/Metagenetics
cd ~/Metagenetics/16s_export
```

Before we start exporting it is best to make a few directories.

```{bash eval=FALSE}
mkdir export
mkdir qiime2_exports
mkdir final_exports
```

- `export`: This directory is where we will point the various `qiime tools export` commands. You can only specify an output directory for this command and the command will give the exported file very generic names. This makes it very easy to overwrite files when using this command.
- `qiime2_exports`: For some of the exported files we will move and rename them from `export` to this directory. This is an temporary directory for files we will not want at the end (mostly BIOM files).
- `final_exports`: This directory will contain the final exported files we would generaly want to keep.

## Abundance table export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_Table_250445_000000.png", auto_pdf = TRUE)
```

First we will export our ASV abundance table. We will have to convert the BIOM file produced by QIIME2 to a TSV file.

Export the table to a directory. This will create a BIOM file called `feature-table.biom`.
```{bash eval=FALSE}
qiime tools export --input-path table-dada2.qza --output-path export
```

Move the created file whilst renaming it.
```{bash eval=FALSE}
mv export/feature-table.biom qiime2_exports/ASV_table.biom
```

Finall we will use `biom convert` to create our TSV file.
```{bash eval=FALSE}
biom convert \
-i qiime2_exports/ASV_table.biom \
-o final_exports/ASV_table.tsv \
--to-tsv
```

The option `--to-tsv` (or alternative) must be provided or the command will not run and only complain.

Use a text viewer/editor of your choice to inspect the final file. Make sure to do this for all the files in this tutorial.

## Taxonomy info export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_ranking_165760.png", auto_pdf = TRUE)
```

Exporting the taxonomy info.

```{bash eval=FALSE}
qiime tools export --input-path taxonomy.sklearn.qza \
--output-path export
#Move the produced tsv file to the final_exports directory
mv export/taxonomy.tsv final_exports
```

## Representative sequences export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_DNA_947999.png", auto_pdf = TRUE)
```

Export the representative sequences to a fasta file.

```{bash eval=FALSE}
qiime tools export --input-path rep-seqs-dada2.qza \
--output-path export
#Move and rename the produced fasta file
mv export/dna-sequences.fasta final_exports/ASV_rep_seqs.fasta
```

## Phylogenetic tree export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_Tree_1857655.png", auto_pdf = TRUE)
```

The below exports the specified phylogenetic tree into a newick format file. For more info on the Newick format please see:  https://en.wikipedia.org/wiki/Newick_format

```{bash eval=FALSE}
qiime tools export --input-path rooted-tree.qza \
--output-path export
#Move and rename the exported newick file
mv export/tree.nwk final_exports/fasttree_rooted_tree.nwk
```

## ASV table with taxonomy export
```{r, fig.align = 'center',out.width= '25%', echo=FALSE }
knitr::include_graphics(path = "figures/table_w_tax.png", auto_pdf = TRUE)
```

This step will produce a TSV file containing the ASV abundance table with the last column containing the taxonomy assignment of the ASV.

To carry out this step you will need the BIOM format produced in the abundance table export and the exported taxonomy file.

We need to change the header of the taxonomy file for it to be compatible with the biom command.

First create a copy which we will edit

```{bash eval=FALSE}
cp final_exports/taxonomy.tsv final_exports/taxonomy.header.tsv
```

Then using a text editor of your choice (nano, vim, emacs, etc.) change the first and second column headers to `#OTUID` and `taxonomy`.

Now we can create a new BIOM file which contains the taxonomy information
```{bash eval=FALSE}
biom add-metadata \
--input-fp qiime2_exports/ASV_table.biom \
--observation-metadata-fp final_exports/taxonomy.header.tsv \
--output-fp qiime2_exports/ASV_table.tax.biom
```

Export the newly created BIOM file. We will need to include option `--header-key taxonomy` to include the taxonomy info in the newly TSV file.
```{bash eval=FALSE}
biom convert \
-i qiime2_exports/ASV_table.tax.biom \
-o final_exports/ASV_table.tax.tsv \
--header-key taxonomy \
--to-tsv
```

## Rarified table export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_curve_3542841.png", auto_pdf = TRUE)
```

You can export a rarefied table. The first step of this is to rarefy our ASV table artifact. 

The option `--p-sampling-depth` is our rarefaction threshold. We will choose 20,000 for this example but make sure to pick an appropriate one for your own projects.

```{bash eval=FALSE}
qiime feature-table rarefy \
--i-table table-dada2.qza \
--p-sampling-depth 20000 \
--o-rarefied-table table-dada2.rarefied_20000.qza
```

After that it is the normal export and convert commands. To be fancy we will include the taxonomy information in the final file.

```{bash eval=FALSE}
#Qiime tools export
qiime tools export \
--input-path table-dada2.rarefied_20000.qza \
--output-path export
#Move and rename exported biom
mv export/feature-table.biom qiime2_exports/ASV_table.rarefied_20000.biom
#Add taxonomy info to biom file
#Ensure to use the taxonomy with the altered headers
biom add-metadata \
--input-fp qiime2_exports/ASV_table.rarefied_20000.biom \
--observation-metadata-fp final_exports/taxonomy.header.tsv \
--output-fp qiime2_exports/ASV_table.rarefied_20000.tax.biom
#Convert biom to tsv with taxonomy
biom convert \
-i qiime2_exports/ASV_table.rarefied_20000.tax.biom \
-o final_exports/ASV_table.rarefied_20000.tax.tsv \
--header-key taxonomy \
--to-tsv
```

## Relative abundance table export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_Percent_3560962.png", auto_pdf = TRUE)
```

Instead of an abundance table you may want a relative abundance table. Thankfully this can easily be done in QIIME 2 with the following command. 

__Note__: You will most likely want to get the relative abundance from a rarefied table.

```{bash eval=FALSE}
qiime feature-table relative-frequency \
--i-table table-dada2.rarefied_20000.qza \
--o-relative-frequency-table table-dada2.rarefied_20000.relabund.qza
```

Then you can export it the same way as the rarefied abundance table.

```{bash eval=FALSE}
#Qiime tools export
qiime tools export \
--input-path table-dada2.rarefied_20000.relabund.qza \
--output-path export
#Move and rename exported biom
mv export/feature-table.biom \
qiime2_exports/ASV_table.rarefied_20000.relabund.biom
#Add taxonomy info to biom file
biom add-metadata \
--input-fp qiime2_exports/ASV_table.rarefied_20000.relabund.biom \
--observation-metadata-fp final_exports/taxonomy.header.tsv \
--output-fp qiime2_exports/ASV_table.rarefied_20000.relabund.tax.biom
#Convert biom to tsv with taxonomy
biom convert \
-i qiime2_exports/ASV_table.rarefied_20000.relabund.tax.biom \
-o final_exports/ASV_table.rarefied_20000.relabund.tax.tsv \
--header-key taxonomy \
--to-tsv
```

## Taxonomy abundance table export
```{r, fig.align = 'center',out.width= '10%', echo=FALSE }
knitr::include_graphics(path = "figures/noun_Family_137805.png", auto_pdf = TRUE)
```

The last recipe I will show you is how to create and then export a taxonomy abundance table.

First we collapse our ASV table to a specific taxonomy level. This is carried out by specifying a number to the option `--p-level` in the command below. With Silva and Greengenes the numbers normally correspond to:

1. Kingdom
2. Phylum
3. Class
4. Order
5. Family
6. Genus
7. Species

Be careful though as with some taxonomy databases the levels are not consistent between different organisms. This occurs if sub-level information (such as sub-order, sub-family etc.) is included in some but not all taxonomic classifications. This could mean that the 5th level taxonomy of one organisms could be its sub-order whilst another's could be its family. Check your `export/taxonomy.tsv` to see if this could be an issue or not.

We will now run our taxonomy collapse command on our rarefied table to produce a Family abundance table.

__Note__: The below command can not be run on a relative abundance table. You will have to run the relative frequency command on its output if you want relative abundance family table.

```{bash eval=FALSE}
qiime taxa collapse \
--i-table table-dada2.rarefied_20000.qza \
--i-taxonomy taxonomy.sklearn.qza \
--p-level 5 \
--o-collapsed-table table-dada2-family.rarefied_20000.qza
```

Then you can export it in a similar way as the other tables. However, we should not include the step that adds the taxonomy for this.

```{bash eval=FALSE}
#Qiime tools export
qiime tools export \
--input-path table-dada2-family.rarefied_20000.qza \
--output-path export
#Move and rename exported biom
mv export/feature-table.biom qiime2_exports/family_table.rarefied_20000.biom
#Convert biom to tsv with taxonomy
biom convert \
-i qiime2_exports/family_table.rarefied_20000.biom \
-o final_exports/family_table.rarefied_20000.tsv \
--to-tsv
```

I hope this is a nice quick reference for your future use.