# Import {#import_chap}
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/import.png", auto_pdf = TRUE)
```

Before carrying out any analysis we first need to import our QIIME2 artifacts into a `phyloseq` object. Thankfully there is an R package called `qiime2R`.

## Import: notebook

```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/r_script.png", auto_pdf = TRUE)
```

Prior to any coding, we will create a new directory, our analysis directory.
afterwards, reate a new notebook called __"01-Import.ipynb"__ in it. 
We will be creating many notebooks, so we'll number them so we can easily see the order of scripts.

First create a new directory.

- In the notebook file explorer, click the __"New"__ button.
- Select __"Folder"__.

You will have an __"Untitled Folder"__. To rename it:

- Click on the box left of the name.
- Press the __"Rename"__ button that appeared.
- Change the name to __"R_community_workshop"__.
- Click __"Rename"__.

Click on your __"R_community_workshop"__ folder to move into it.

Next step is to create a new R notebook, rename it to __"01-Import"__, and save it.

## qiime2R

```{r, fig.align = 'center',out.width= '40%', echo=FALSE }
knitr::include_graphics(path = "figures/qiime2r.png", auto_pdf = TRUE)
```

[`qiime2R`](https://github.com/jbisanz/qiime2R) is an R package for importing QIIME2 artifacts into a R `phyloseq` object. The package contains many different commands. Its function `read_qza()` can read a single artifact at a time.

The best way to import all your QIIME2 artifacts is with the `qiime2R::qza_to_phyloseq()` function. In your __"01-Import.ipynb"__ script, add the following and run the commands.

__Tip__: You can tab complete and/or copy and paste file paths within the webVNC.

```{r, eval=FALSE}
#Cell 1
#Load the package/library
library("qiime2R")

#Import data
pseq <- qiime2R::qza_to_phyloseq(
  features = "/pub14/tea/nsc206/NEOF/R_community/data/table-dada2.qza",
  tree = "/pub14/tea/nsc206/NEOF/R_community/data/rooted-tree.qza",
  taxonomy = "/pub14/tea/nsc206/NEOF/R_community/data/taxonomy.sklearn.qza",
  metadata = "/pub14/tea/nsc206/NEOF/R_community/data/media_metadata.txt"
)
```

This command creates a phyloseq object named `pseq`. It contains:

-   The ASV abundance table (`features = "table-dada2.qza"`).
-   The rooted phylogenetic tree (`tree = "rooted-tree.qza"`).
-   The taxonomic classifications of the ASVs (`taxonomy = "taxonomy.sklearn.qza"`).
-   The sample metadata (`metadata = "media_metadata.txt"`)

## Import: Summarise phyloseq
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/list_blue.png", auto_pdf = TRUE)
```

Now that we have imported the data we can extract some summary information from it.

First we will use the `microbiome` package with its `summarize_phyloseq()` function.

Create a new __cell__ and write + run the below in it.

```{r, eval=FALSE}
#Cell 2
#Load microbiome library
library("microbiome")
#Summary of phyloseq object
microbiome::summarize_phyloseq(pseq)
```

This gives us a plethora of information:

-   The top line tells us if the data is compositional (relative abundance).
-   We get the following list of values in a paragraph and via a list.
    -   __Min. number of reads:__ Number of reads in the sample with the lowest number of reads.
    -   __Max. number of reads:__ Number of reads in the sample with the largest number of reads.
    -   __Total number of reads:__ Sum of all reads across all samples.
    -   __Average number of reads:__ Sum of all reads / number of samples.
    -   __Median number of reads:__ Midpoint read abundance across samples.
    -   __Sparsity:__ See expandable box further down.
    -   __Any OTU sum to 1 or less?:__ States if there are any ASVs with a summed abundance of 1 or less across all the samples. 
    -   __Number of singletons:__ Number of ASVs with a sum of 1 across all samples.
    -   __Percent of OTUs that are singletons:__ Percentage of ASVs that only contain one read across all the samples.
    -   __Number of sample variables are:__ Number of sample variables/groupings in our metadata.
    -   The last line shows the names of the sample variables/groupings in our metadata.

`r hide("Sparsity")`
Sparsity is a measure of the number of 0s in a table. It can be represented by the following equation:

$$
sparsity = Z/C
$$

Where:

-   Z = The number of cells that equal zero.
-   C = The total number of cells.

Let's look at an example of an abundance table with a small amount of ASVs and Samples.

|      | Sample1 | Sample2 | Sample3 |
|------|---------|---------|---------|
| ASV1 | 0       | 10      | 24      |
| ASV2 | 1       | 0       | 37      |
| ASV3 | 6       | 25      | 0       |
| ASV4 | 51      | 2       | 0       |

- This abundance table has 12 cells (3 samples \* 4 ASVs). 
- Of these cells, 4 have an abundance of zero. 
- 4/12 = 0.3333, therefore its sparsity is 0.3333.

Sparsity can be any value from 0-1. The higher the value the more sparse it is, with a value of 1 meaning all the cells have an abundance of zero. The lower the value the less sparse it is, with a value of 0 meaning all the cells have an abundance of 1 or more.

16S data is known to be sparse so high sparsity is not unexpected. Keep in mind that lower levels of taxa (ASVs, Species, & Genera) will generally have more sparse tables that higher levels of taxa (Kingdom, Phylum, Class).
`r unhide()`

If you would like to see how the function calculates its values you can view the [source code online](https://rdrr.io/github/microbiome/microbiome/src/R/summarize_phyloseq.R).

## Save the phyloseq object
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/r_save.png", auto_pdf = TRUE)
```

When using multiple notebooks/scripts for analysis it is useful to save the R objects that will be used in different notebooks/scripts. This can be carried out with the function `save()`.

Write and run the following code in a third __cell__.

```{r, eval=FALSE}
#Cell 3
#Save phyloseq as file
save(pseq, file = "phyloseq.RData")
```

This saves our object `pseq` into the file `phyloseq.RData`. The suffix `.RData` is the normal convention.

We have saved our final object of the notebook. Close and halt it.

## Import: recap
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/recap.png", auto_pdf = TRUE)
```

We have imported our QIIME2 artifacts into one `phyloseq` object so we can analyse the data in R. This object has been saved into a ".RData" file which we will load in the next chapter.
