# Preprocessing
```{r, fig.align = 'center',out.width= '20%', echo=FALSE }
knitr::include_graphics(path = "figures/transform.png", auto_pdf = TRUE)
```

In this chapter we will learn how to transform our abundance phyloseq table into a relative/compositional abundance phyloseq and a rarefied phyloseq object.

## Setup
```{r, fig.align = 'center',out.width= '15%', echo=FALSE }
knitr::include_graphics(path = "figures/setup_2.png", auto_pdf = TRUE)
```

Before starting analysis create and save a new `R` notebook called __"02-Preprocess.ipynb"__ in the analysis directory (when renaming you don't need to include the suffix).

It is useful to add a title to the top of the first cell. Add the below to the cell:

```{r, eval=FALSE}
#Preprocessing data notebook
```

We will use our first cell to carry out setup. This will involve loading libraries and data we need. 

For good documentation add a title for our cell.

```{r, eval=FALSE}
#Setup cell
```

I like to load all the libraries to be used in the notebook in this section. We will explain their uses later in this chapter.

Add the below to your notebook:

```{r, eval=FALSE}
#Libraries
library("phyloseq")
library("microbiome")
library("vegan")
```

Our last bit of set-up is to load in our abundance phyloseq object we created in the previous chapter.

```{r, eval=FALSE}
#Load the phyloseq object
load("phyloseq.RData")
```

Ensure you have run the code in this cell.

## Preprocess: Summarise 

When you load in a dataset it is always useful to check it. We will therefore use new cells and to inspect the data.

### Summarise phyloseq

Create a second cell and add the following annotation and code. Then run the code.

```{r, eval=FALSE}
#Summarise the phyloseq object
#Summary of phyloseq object
microbiome::summarize_phyloseq(pseq)
```

We ran this code in the __"01-Import.ipynb"__. You can therefore check if our new output matched the output from that notebook. This should be the case since they are the same data.

Due to the relative large amount of output to screen we'll put the next part into a new cell (third cell).

### Reads per sample

In our third cell we will use the command `microbiome::readcount()` to store and display the number of reads in each sample within our `phyloseq` object (`pseq`). We can even make a quick histogram of read numbers per sample with the base `R` function `hist()`.

Write and run the below code in a third cell.

```{r, eval=FALSE}
#Number of reads per sample
reads_sample <- microbiome::readcount(pseq)
reads_sample
#Histogram
hist(reads_samples, "Histogram of read depths")
```

With this information we can decide if our samples have enough depth or not. Normally, for 16S data, a depth of at least 20K per sample is suggested. However, this is the general consensus for human microbiome data. 

There are 2 main considerations to take into account for what is an appropriate depth for your dataset.

- The biodiversity of your sample.
  - If your sample is very biodiverse, such as the human gut microbiome, you will need a good depth (>20K per sample).
  - If your sample is less biodiverse, such as many geological environments or skin, then you will not need as much read depth.
  - Rarefaction curves are a good method to determine if your samples have enough depth. We will look at this later in this chapter.
- The biomass of your sample.
  - Some environments are hard to extract DNA from.
  - If this is the case for you then people will hopefully accept that this is an unfortunate reality of life and you will use what you can.
  - However, be careful of your conclusions, if you think your data doesn't have as much as it could do not make very definitive detailed claims.

That is a brief overview of that topic. If you are interested in more I suggest you look at papers where they have studied an environment similar to yours.

### ASVs per sample

The last feature we will look at before some preprocessing is the ASVs (Amplicon Feature Variants).

#Can extract ASV table (known as otu table in phyloseq)
phyloseq::otu_table(pseq)

#Each row is an ASV and each column is the samples
#Therefore we can get the number of ASVs in data
#Let's make a new vector with this info so we can easily keep track
num_asvs_vec <- c(nrow(phyloseq::otu_table(pseq)))
#Give the 1st element a relevant name
names(num_asvs_vec)[1] <- "abundance"
#View current vector
num_asvs_vec

#Sample with too few samples?
#In your analysis you may have a sample with too few reads
#All our samples are fine for a tutorial case but let us say we only wanted to
#keep samples with >11k reads
#We could remove samples with the following code
pseq_min11K <- phyloseq::subset_samples(pseq, reads_sample > 11000)
microbiome::summarize_phyloseq(pseq_min11K)
microbiome::readcount(pseq_min11K)

#We won't be using this as we are happy with our sample numbers
#Let us therefore remove this sample subsetted variable
rm(pseq_min11K)

#Relative abundance ####
#Convert abundance table to relative abundance (compositional) table
pseq_relabund <- microbiome::transform(pseq, "compositional")
#Summarise and check sample counts which should each amount to 1
microbiome::summarize_phyloseq(pseq_relabund)
microbiome::readcount(pseq_relabund)

#Check the below logic
#When using total abundance values it is useful to have 0 values, singletons, and doubletons
#This is because some alpha diversity metrics require them
#However it is useful to remove low relative abundance data in relative abundance data
#This is so the rare ASVs do not overly affect certain types of analysis

#first remove ASV with relabund equal to 0
#This can occur if samples were removed which had ASVs 
#not present in the remaining samples
pseq_relabund <-  filter_taxa(pseq_relabund, function(x) sum(x) > 0, TRUE)

#Summarise and check sample counts which should each amount to around 1
microbiome::summarize_phyloseq(pseq_relabund)
microbiome::readcount(pseq_relabund)

#All the total relative abundances still equal 1
#This is expected since no samples were removed
#As this is the case there is no need to check if ASVs were removed

#We will now remove rare ASVs as these are not as useful in relative abundance data
#compared to abundance data
#There are many ways to do this
#A common way, recommended by the phyloseq developer
#Remove ASVs with a mean (across samples) less than 1e-5 (relabund)
pseq_relabund <-  
  phyloseq::filter_taxa(
    pseq_relabund, function(x) mean(x) > 1e-5, TRUE)

#Summarise and check sample counts which should each amount to around 1
microbiome::summarize_phyloseq(pseq_relabund)
microbiome::readcount(pseq_relabund)

#Total relative abundance has decreased by a very small amount
#This is what we are looking for, if too much is being removed >0.05
#you will need to try to be gentler with the filtering
#Such as trying 1e-6 rather than 1e-5

#We should also check how many ASVs have been removed
num_asvs_vec["relabund"] <- nrow(phyloseq::otu_table(pseq_relabund))
num_asvs_vec
num_asvs_vec["abundance"] - num_asvs_vec["relabund"]

#We have lost a good amount of ASVs but these only equate to a very small
#amount relabund. This is fine as we generally use relative abundance
#when looking at the larger picture rather than at closer pictures
#instead we can use a rarefied abundance table to look at the closer picture

#We are now happy with our relative abundance table
#Therefore we can save it for further use
save(pseq_relabund, file = "phyloseq_relabund.RData")

#Rarefy abundance table ####
#i.e. convert abundance numbers so each sample has equal depth

#Before rarefying a table it is good to make a rarefaction curve
#This is to help us choose an appropriate rarefaction threshold

#We will use the very useful package vegan
#Ignore any warning message
vegan::rarecurve(
  x = as.data.frame(t(phyloseq::otu_table(pseq))), step = 50)

#Let us improve this and save it into a file
png(filename = "./rarefaction_plot.png", res = 300,
    units = "mm", height = 200, width = 300)
vegan::rarecurve(
  x = as.data.frame(t(phyloseq::otu_table(pseq))), step = 50,
  ylab="ASVs", lwd=1,label=F)
#Add a vertical line of the smallest sample depth
abline(v = min(reads_sample), col="red")
dev.off()

#With this we can see that the majorty of samples plateau at 
#the minimum sampleing depth
#Therefore we can use this as a rarefaction size
pseq_rarefy <- 
  phyloseq::rarefy_even_depth(
    pseq, sample.size = min(reads_sample), rngseed = 1000)

#Summarise and check sample counts which should each amount to 10433
microbiome::summarize_phyloseq(pseq_rarefy)
microbiome::readcount(pseq_rarefy)

#Check ASVs
num_asvs_vec["rarefied"] <- nrow(phyloseq::otu_table(pseq_rarefy))
num_asvs_vec

#Save phyloseq object
save(pseq_relpseq_rarefyabund, file = "phyloseq_rarefied.RData")
